
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="../_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="../_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#2196f3">
  <script src="../_static/javascripts/modernizr.js"></script>
  
    <script async src="../_static/cookieconsent.min.js"></script>
    <script>
        window.addEventListener("load", function(){
        window.cookieconsent.initialise({
            "palette": {
            "popup": {
                "background": "#f0f0f0",
                "text": "#999"
            },
            "button": {
                "text": "#fff",
                "background": "#009688"
            }
            },
            "theme": "classic"
        })});
    </script>
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-Q6ZE82CNZB"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'G-Q6ZE82CNZB');
    </script>
  
    <link rel="apple-touch-icon" href="../_static/images/apple-icon-152x152.png"/>
  
  
    <title>T010 ¬∑ Binding site similarity and off-target prediction &#8212; TeachOpenCADD 0 documentation</title>
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/material.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Complete list of talktorials" href="../all_talktorials.html" />
    <link rel="prev" title="T009 ¬∑ Ligand-based pharmacophores" href="T009_compound_ensemble_pharmacophores.html" />
  
    <link rel="apple-touch-icon" href="../_static/images/apple-icon-152x152.png"/>
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=teal data-md-color-accent=cyan>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#talktorials/T010_binding_site_comparison" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../index.html" title="TeachOpenCADD 0 documentation"
           class="md-header-nav__button md-logo">
          
            <i class="md-icon">school</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">TeachOpenCADD (WIP üöß)</span>
          <span class="md-header-nav__topic"> T010 ¬∑ Binding site similarity and off-target prediction </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="../search.html" method="GET" name="search">
      <input type="text" class="md-search__input" name="q" placeholder="Search"
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            <a href="https://github.com/volkamerlab/teachopencadd/" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    TeachOpenCADD
  </div>
</a>
          </div>
        </div>
      
      
    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
            
            <li class="md-tabs__item"><a href="../talktorials.html" class="md-tabs__link">Our talktorials</a></li>
            
            <li class="md-tabs__item"><a href="../installing.html" class="md-tabs__link">Run locally</a></li>
            
            <li class="md-tabs__item"><a href="../contribute.html" class="md-tabs__link">Contribute</a></li>
          <li class="md-tabs__item"><a href="../talktorials.html" class="md-tabs__link">Talktorials by collection</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../index.html" title="TeachOpenCADD 0 documentation" class="md-nav__button md-logo">
      
        <i class="md-icon">school</i>
      
    </a>
    <a href="../index.html"
       title="TeachOpenCADD 0 documentation">TeachOpenCADD (WIP üöß)</a>
  </label>
    <div class="md-nav__source">
      <a href="https://github.com/volkamerlab/teachopencadd/" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    TeachOpenCADD
  </div>
</a>
    </div>
  
  

  
  <ul class="md-nav__list">
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Our talktorials</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../talktorials.html" class="md-nav__link">Talktorials by collection</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../all_talktorials.html" class="md-nav__link">Complete list of talktorials</a>
      <ul class="md-nav__list"> 
    <li class="md-nav__item">
    
    
      <a href="T001_query_chembl.html" class="md-nav__link">T001 ¬∑ Compound data acquisition (ChEMBL)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T002_compound_adme.html" class="md-nav__link">T002 ¬∑ Molecular filtering: ADME and lead-likeness criteria</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T003_compound_unwanted_substructures.html" class="md-nav__link">T003 ¬∑ Molecular filtering: unwanted substructures</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T004_compound_similarity.html" class="md-nav__link">T004 ¬∑ Ligand-based screening: compound similarity</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T005_compound_clustering.html" class="md-nav__link">T005 ¬∑ Compound clustering</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T006_compound_maximum_common_substructures.html" class="md-nav__link">T006 ¬∑ Maximum common substructure</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T007_compound_activity_machine_learning.html" class="md-nav__link">T007 ¬∑ Ligand-based screening: machine learning</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T008_query_pdb.html" class="md-nav__link">T008 ¬∑ Protein data acquisition: Protein Data Bank (PDB)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T009_compound_ensemble_pharmacophores.html" class="md-nav__link">T009 ¬∑ Ligand-based pharmacophores</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    <label class="md-nav__link md-nav__link--active" for="__toc"> T010 ¬∑ Binding site similarity and off-target prediction </label>
    
      <a href="#" class="md-nav__link md-nav__link--active">T010 ¬∑ Binding site similarity and off-target prediction</a>
      
        
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#talktorials-t010-binding-site-comparison--page-root" class="md-nav__link">T010 ¬∑ Binding site similarity and off-target prediction</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Aim-of-this-talktorial" class="md-nav__link">Aim of this talktorial</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Contents-in-Theory" class="md-nav__link">Contents in <em>Theory</em></a>
        </li>
        <li class="md-nav__item"><a href="#Contents-in-Practical" class="md-nav__link">Contents in <em>Practical</em></a>
        </li>
        <li class="md-nav__item"><a href="#References" class="md-nav__link">References</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Theory" class="md-nav__link">Theory</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Off-target-proteins" class="md-nav__link">Off-target proteins</a>
        </li>
        <li class="md-nav__item"><a href="#Computational-off-target-prediction:-binding-site-comparison" class="md-nav__link">Computational off-target prediction: binding site comparison</a>
        </li>
        <li class="md-nav__item"><a href="#Pairwise-RMSD-as-simple-measure-for-similarity" class="md-nav__link">Pairwise RMSD as simple measure for similarity</a>
        </li>
        <li class="md-nav__item"><a href="#Imatinib,-a-tyrosine-kinase-inhibitor" class="md-nav__link">Imatinib, a tyrosine kinase inhibitor</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Practical" class="md-nav__link">Practical</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Load-and-visualize-the-ligand-of-interest-(Imatinib/STI)" class="md-nav__link">Load and visualize the ligand of interest (Imatinib/STI)</a>
        </li>
        <li class="md-nav__item"><a href="#Get-all-protein-STI-complexes-from-the-PDB" class="md-nav__link">Get all protein-STI complexes from the PDB</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Query-the-PDB" class="md-nav__link">Query the PDB</a>
        </li>
        <li class="md-nav__item"><a href="#Filter-the-PDB-data-set" class="md-nav__link">Filter the PDB data set</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Visualize-the-PDB-structures" class="md-nav__link">Visualize the PDB structures</a>
        </li>
        <li class="md-nav__item"><a href="#Align-the-PDB-structures-(full-protein)" class="md-nav__link">Align the PDB structures (full protein)</a>
        </li>
        <li class="md-nav__item"><a href="#Get-pairwise-RMSD-(full-protein)" class="md-nav__link">Get pairwise RMSD (full protein)</a>
        </li>
        <li class="md-nav__item"><a href="#Align-PDB-structures-(binding-sites)" class="md-nav__link">Align PDB structures (binding sites)</a>
        </li>
        <li class="md-nav__item"><a href="#Get-pairwise-RMSD-(binding-sites)" class="md-nav__link">Get pairwise RMSD (binding sites)</a>
        </li>
        <li class="md-nav__item"><a href="#Filter-out-outliers" class="md-nav__link">Filter out outliers</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Discussion" class="md-nav__link">Discussion</a>
        </li>
        <li class="md-nav__item"><a href="#Quiz" class="md-nav__link">Quiz</a>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/talktorials/T010_binding_site_comparison.nblink.txt">Show Source</a> </li>

  </ul>
</nav>
      
    
    </li></ul>
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Run locally</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../installing.html" class="md-nav__link">Installing</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Contributors</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../contribute.html" class="md-nav__link">For contributors</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../api.html" class="md-nav__link">API Documentation</a>
      
    
    </li>
  </ul>
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#talktorials-t010-binding-site-comparison--page-root" class="md-nav__link">T010 ¬∑ Binding site similarity and off-target prediction</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Aim-of-this-talktorial" class="md-nav__link">Aim of this talktorial</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Contents-in-Theory" class="md-nav__link">Contents in <em>Theory</em></a>
        </li>
        <li class="md-nav__item"><a href="#Contents-in-Practical" class="md-nav__link">Contents in <em>Practical</em></a>
        </li>
        <li class="md-nav__item"><a href="#References" class="md-nav__link">References</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Theory" class="md-nav__link">Theory</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Off-target-proteins" class="md-nav__link">Off-target proteins</a>
        </li>
        <li class="md-nav__item"><a href="#Computational-off-target-prediction:-binding-site-comparison" class="md-nav__link">Computational off-target prediction: binding site comparison</a>
        </li>
        <li class="md-nav__item"><a href="#Pairwise-RMSD-as-simple-measure-for-similarity" class="md-nav__link">Pairwise RMSD as simple measure for similarity</a>
        </li>
        <li class="md-nav__item"><a href="#Imatinib,-a-tyrosine-kinase-inhibitor" class="md-nav__link">Imatinib, a tyrosine kinase inhibitor</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Practical" class="md-nav__link">Practical</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Load-and-visualize-the-ligand-of-interest-(Imatinib/STI)" class="md-nav__link">Load and visualize the ligand of interest (Imatinib/STI)</a>
        </li>
        <li class="md-nav__item"><a href="#Get-all-protein-STI-complexes-from-the-PDB" class="md-nav__link">Get all protein-STI complexes from the PDB</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Query-the-PDB" class="md-nav__link">Query the PDB</a>
        </li>
        <li class="md-nav__item"><a href="#Filter-the-PDB-data-set" class="md-nav__link">Filter the PDB data set</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Visualize-the-PDB-structures" class="md-nav__link">Visualize the PDB structures</a>
        </li>
        <li class="md-nav__item"><a href="#Align-the-PDB-structures-(full-protein)" class="md-nav__link">Align the PDB structures (full protein)</a>
        </li>
        <li class="md-nav__item"><a href="#Get-pairwise-RMSD-(full-protein)" class="md-nav__link">Get pairwise RMSD (full protein)</a>
        </li>
        <li class="md-nav__item"><a href="#Align-PDB-structures-(binding-sites)" class="md-nav__link">Align PDB structures (binding sites)</a>
        </li>
        <li class="md-nav__item"><a href="#Get-pairwise-RMSD-(binding-sites)" class="md-nav__link">Get pairwise RMSD (binding sites)</a>
        </li>
        <li class="md-nav__item"><a href="#Filter-out-outliers" class="md-nav__link">Filter out outliers</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Discussion" class="md-nav__link">Discussion</a>
        </li>
        <li class="md-nav__item"><a href="#Quiz" class="md-nav__link">Quiz</a>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/talktorials/T010_binding_site_comparison.nblink.txt">Show Source</a> </li>

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>

<h1 id="talktorials-t010-binding-site-comparison--page-root">T010 ¬∑ Binding site similarity and off-target prediction<a class="headerlink" href="#talktorials-t010-binding-site-comparison--page-root" title="Permalink to this headline">¬∂</a></h1>
<p>Authors:</p>
<ul class="simple">
<li><p>Angelika Szengel, CADD seminar 2017, Charit√©/FU Berlin</p></li>
<li><p>Marvis Sydow, CADD seminar 2018, Charit√©/FU Berlin</p></li>
<li><p>Richard Gowers, RDKit UGM hackathon 2019</p></li>
<li><p>Jaime Rodr√≠guez-Guerra, 2020, <a class="reference external" href="https://volkamerlab.org">Volkamer lab</a>, Charit√©</p></li>
<li><p>Dominique Sydow, 2018-2020, <a class="reference external" href="https://volkamerlab.org">Volkamer lab</a>, Charit√©</p></li>
<li><p>Mareike Leja, 2020, <a class="reference external" href="https://volkamerlab.org">Volkamer lab</a>, Charit√©</p></li>
</ul>
<p><strong>Talktorial T010</strong>: This talktorial is part of the TeachOpenCADD pipeline described in the <a class="reference external" href="https://jcheminf.biomedcentral.com/articles/10.1186/s13321-019-0351-x">first TeachOpenCADD paper</a>, comprising of talktorials T001-T010.</p>
<p><strong>Note</strong>: Please run this notebook cell by cell. Running all cells in one is possible also, however, part of the <code class="docutils literal notranslate"><span class="pre">nglview</span></code> 3D representations might be missing.</p>

<h2 id="Aim-of-this-talktorial">Aim of this talktorial<a class="headerlink" href="#Aim-of-this-talktorial" title="Permalink to this headline">¬∂</a></h2>
<p>In this talktorial, we use the structural similarity of full proteins and binding sites to predict off-targets, i.e.¬†proteins that are not intended targets of a drug. This may lead to unwanted side effects or enable desired alternate applications of a drug (drug repositioning). We discuss the main steps for binding site comparison and implement a basic method, i.e.¬†the geometrical variation between structures (the root mean square deviation between two structures).</p>

<h3 id="Contents-in-Theory">Contents in <em>Theory</em><a class="headerlink" href="#Contents-in-Theory" title="Permalink to this headline">¬∂</a></h3>
<ul class="simple">
<li><p>Off-target proteins</p></li>
<li><p>Computational off-target prediction: binding site comparison</p></li>
<li><p>Pairwise RMSD as simple measure for similarity</p></li>
<li><p>Imatinib, a tyrosine kinase inhibitor</p></li>
</ul>


<h3 id="Contents-in-Practical">Contents in <em>Practical</em><a class="headerlink" href="#Contents-in-Practical" title="Permalink to this headline">¬∂</a></h3>
<ul class="simple">
<li><p>Load and visualize the ligand of interest (Imatinib/STI)</p></li>
<li><p>Get all protein-STI complexes from the PDB</p>
<ul>
<li><p>Query the PDB</p></li>
<li><p>Filter the PDB data set</p></li>
</ul>
</li>
<li><p>Visualize the PDB structures</p></li>
<li><p>Align the PDB structures (full protein)</p></li>
<li><p>Get pairwise RMSD (full protein)</p></li>
<li><p>Align the PDB structures (binding site)</p></li>
<li><p>Get pairwise RMSD (binding site)</p></li>
<li><p>Filter out outliers</p></li>
</ul>


<h3 id="References">References<a class="headerlink" href="#References" title="Permalink to this headline">¬∂</a></h3>
<p>Binding site superposition + comparison:</p>
<ul class="simple">
<li><p>Binding site comparison reviews:</p>
<ul>
<li><p><a class="reference external" href="https://www.eurekaselect.com/67606/article/how-measure-similarity-between-protein-ligand-binding-sites">Curr. Comput. Aided Drug Des. (2008), 4, 209-20</a></p></li>
<li><p><a class="reference external" href="https://pubs.acs.org/doi/10.1021/acs.jmedchem.6b00078">J. Med. Chem. (2016), 9, 4121-51</a></p></li>
</ul>
</li>
<li><p>Molecular superposition with Python: <code class="docutils literal notranslate"><span class="pre">opencadd</span></code> package (<code class="docutils literal notranslate"><span class="pre">structure.superposition</span></code> module) (<a class="reference external" href="https://github.com/volkamerlab/opencadd">GitHub repository</a>)</p></li>
<li><p>Wikipedia article on root mean square deviation (<a class="reference external" href="https://en.wikipedia.org/wiki/Root-mean-square_deviation_of_atomic_positions">RMSD</a>) and <a class="reference external" href="https://en.wikipedia.org/wiki/Structural_alignment">structural superposition</a></p></li>
<li><p>Structural superposition: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/B9780128001684000056?via%3Dihub">Book chapter: Algorithms, Applications, and Challenges of Protein Structure Alignment in Advances in Protein Chemistry and Structural Biology (2014), 94, 121-75</a></p></li>
</ul>
<p>Imatinib:</p>
<ul class="simple">
<li><p>Review on Imatinib: <a class="reference external" href="https://www.nature.com/articles/nrclinonc.2016.41">Nat. Rev.¬†Clin. Oncol. (2016), 13, 431-46</a></p></li>
<li><p>Promiscuity of imatinib: <a class="reference external" href="https://jbiol.biomedcentral.com/articles/10.1186/jbiol134">J. Biol. (2009), 8, 30</a></p></li>
<li><p><a class="reference external" href="https://www.ebi.ac.uk/chembl/compound/inspect/CHEMBL941">ChEMBL information on Imatinib</a></p></li>
<li><p><a class="reference external" href="https://www3.rcsb.org/ligand/STI">PDB information on Imatinib</a></p></li>
<li><p>Side effects of Imatinib</p>
<ul>
<li><p><a class="reference external" href="https://www.drugs.com/cdi/imatinib.html">MedFacts Consumer Drug Information</a></p></li>
<li><p><a class="reference external" href="https://go.drugbank.com/drugs/DB00619">DrugBank</a></p></li>
<li><p><a class="reference external" href="https://bmcstructbiol.biomedcentral.com/articles/10.1186/1472-6807-9-7">BMC Struct. Biol. (2009), 9</a></p></li>
</ul>
</li>
</ul>



<h2 id="Theory">Theory<a class="headerlink" href="#Theory" title="Permalink to this headline">¬∂</a></h2>

<h3 id="Off-target-proteins">Off-target proteins<a class="headerlink" href="#Off-target-proteins" title="Permalink to this headline">¬∂</a></h3>
<p>An off-target can be any protein which interacts with a drug or (one of) its metabolite(s) without being the designated target protein. The molecular reaction caused by the off-target can lead to unwanted side effects, ranging from a rather harmless to an extremely harmful impact. Off-targets mainly occur because on- and off-targets share similar structural motifs with each other in their binding site and therefore can bind similar ligands.</p>


<h3 id="Computational-off-target-prediction:-binding-site-comparison">Computational off-target prediction: binding site comparison<a class="headerlink" href="#Computational-off-target-prediction:-binding-site-comparison" title="Permalink to this headline">¬∂</a></h3>
<p>Computation-aided prediction of potential off-targets is aimed at minimizing the risk of developing potentially dangerous substances for medical treatment. There are several algorithmic approaches to assess binding site similarity but they always consist of three main steps:</p>
<ol class="arabic simple">
<li><p><strong>Binding site encoding</strong>: binding sites are encoded using different descriptor techniques and stored in a target database.</p></li>
<li><p><strong>Binding site comparison</strong>: a query binding site is compared with the target database, using different similarity measures.</p></li>
<li><p><strong>Target ranking</strong>: targets are ranked based on a suitable scoring approach.</p></li>
</ol>
<p>For detailed information on different similarity measures and existing tools, we refer to two excellent reviews on binding site comparison (<a class="reference external" href="https://www.eurekaselect.com/67606/article/how-measure-similarity-between-protein-ligand-binding-sites">Curr. Comput. Aided Drug Des. (2008), 4, 209-20</a> and <a class="reference external" href="https://pubs.acs.org/doi/10.1021/acs.jmedchem.6b00078">J. Med. Chem. (2016), 9, 4121-51</a>).</p>
<img alt="Binding site comparison steps" src="../_images/binding_site_comparison_steps.png"/>
<p><em>Figure 1</em>: Main steps of binding site comparison methods (figure by Dominique Sydow).</p>


<h3 id="Pairwise-RMSD-as-simple-measure-for-similarity">Pairwise RMSD as simple measure for similarity<a class="headerlink" href="#Pairwise-RMSD-as-simple-measure-for-similarity" title="Permalink to this headline">¬∂</a></h3>
<p>A simple and straightforward method for scoring the similarity is to use the calculated root mean square deviation (RMSD), which is the square root of the mean of the square of the distances between the atoms of two aligned structures (<a class="reference external" href="https://en.wikipedia.org/wiki/Root-mean-square_deviation_of_atomic_positions">Wikipedia</a>).</p>
<p>In order to find the respective atoms that are compared between two structures, they need to be aligned first based on sequence-based or sequence-independent alignment algorithms (<a class="reference external" href="https://www.sciencedirect.com/science/article/pii/B9780128001684000056?via%3Dihub">Book chapter: Algorithms, Applications, and Challenges of Protein Structure Alignment in Advances in Protein Chemistry and Structural Biology (2014), 94, 121-75</a>).</p>


<h3 id="Imatinib,-a-tyrosine-kinase-inhibitor">Imatinib, a tyrosine kinase inhibitor<a class="headerlink" href="#Imatinib,-a-tyrosine-kinase-inhibitor" title="Permalink to this headline">¬∂</a></h3>
<p>Kinases transfer a phosphate group from ATP to proteins, and thereby regulate various cellular processes such as signal transduction, metabolism, and protein regulation. If these kinases are constitutively active (due to genomic mutations), they can distort regulation processes and cause cancer. An example for cancer treatment is Imatinib (<a class="reference external" href="https://www.nature.com/articles/nrclinonc.2016.41">Nat. Rev.¬†Clin. Oncol. (2016), 13, 431-46</a>), a small molecule tyrosine kinase inhibitor used to treat
cancer, more specifically chronic myeloid leukaemia (CML) and gastrointestinal stromal tumour (GIST).</p>
<p>Imatinib was shown to be not entirely specific and to target tyrosine kinases other than its main target. This was used for drug repositioning, i.e.¬†Imatinib was approved for alternate cancer types, (<a class="reference external" href="https://jbiol.biomedcentral.com/articles/10.1186/jbiol134">J. Biol. (2009), 8, 30</a>), however can also show unwanted side effects such as signs of an allergic reaction, infection, bleeding, or headache (<a class="reference external" href="https://www.drugs.com/cdi/imatinib.html">MedFacts Consumer Drug Information</a>).</p>



<h2 id="Practical">Practical<a class="headerlink" href="#Practical" title="Permalink to this headline">¬∂</a></h2>
<p>In the following, we will fetch and filter PDB structures that bind Imatinib. We will investigate the structure similarity of Imatinib-binding proteins (those with a solved protein structure). The similarity measure used is a pairwise RMSD calculation (as a simple similarity measure), in order to show that this simple method can be used as an initial test for potential off-targets.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">redo</span>
<span class="kn">import</span> <span class="nn">nglview</span> <span class="k">as</span> <span class="nn">nv</span>
<span class="kn">import</span> <span class="nn">pypdb</span>
<span class="kn">from</span> <span class="nn">rdkit</span> <span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span> <span class="nn">rdkit.Chem</span> <span class="kn">import</span> <span class="n">Draw</span><span class="p">,</span> <span class="n">AllChem</span>
<span class="kn">from</span> <span class="nn">MDAnalysis.analysis</span> <span class="kn">import</span> <span class="n">rms</span>
<span class="kn">from</span> <span class="nn">opencadd.structure.core</span> <span class="kn">import</span> <span class="n">Structure</span>
<span class="kn">from</span> <span class="nn">opencadd.structure.superposition.api</span> <span class="kn">import</span> <span class="n">align</span><span class="p">,</span> <span class="n">METHODS</span>
<span class="kn">from</span> <span class="nn">opencadd.structure.superposition.engines.mda</span> <span class="kn">import</span> <span class="n">MDAnalysisAligner</span>

<span class="kn">from</span> <span class="nn">teachopencadd.utils</span> <span class="kn">import</span> <span class="n">seed_everything</span>

<span class="n">seed_everything</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "fabbe69cc08b4aee9cde4cce6332b6de", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Ignore warnings, added because of the two following warnings</span>
<span class="c1"># VisibleDeprecationWarning (MDAnalysis), ClusterWarning (seaborn)</span>
<span class="c1"># TODO check in the future if ignoring warnings is still necessary</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">"ignore"</span><span class="p">)</span>
<span class="c1"># Also make MDAnalysis code in OpenCADD a bit less noisy</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">"opencadd"</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Set path to this notebook</span>
<span class="n">HERE</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">_dh</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">DATA</span> <span class="o">=</span> <span class="n">HERE</span> <span class="o">/</span> <span class="s2">"data"</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Frozen set of PDB IDs that will be used in this notebook</span>
<span class="c1"># TODO check in the future if we want to update this dataset</span>
<span class="n">FROZEN_PDB_IDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">"3HEC"</span><span class="p">,</span> <span class="s2">"2PL0"</span><span class="p">,</span> <span class="s2">"4CSV"</span><span class="p">,</span> <span class="s2">"4R7I"</span><span class="p">,</span> <span class="s2">"1XBB"</span><span class="p">,</span> <span class="s2">"3FW1"</span><span class="p">,</span> <span class="s2">"1T46"</span><span class="p">]</span>
</pre></div>
</div>
</div>

<h3 id="Load-and-visualize-the-ligand-of-interest-(Imatinib/STI)">Load and visualize the ligand of interest (Imatinib/STI)<a class="headerlink" href="#Load-and-visualize-the-ligand-of-interest-(Imatinib/STI)" title="Permalink to this headline">¬∂</a></h3>
<p>The SMILES format for Imatinib (common abbreviation: STI) can be retrieved from e.g.¬†the ChEMBL database (<a class="reference external" href="https://www.ebi.ac.uk/chembl/compound/inspect/CHEMBL941">ChEMBL: Imatinib</a>) or the PDB database by its common abbreviation STI (<a class="reference external" href="https://www3.rcsb.org/ligand/STI">PDB: STI</a>). We simply copy the string from the ‚ÄúIsomeric SMILES‚Äù entry of the Chemical Component Summary table, and load the ligand here by hand.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">smiles</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="s2">"CN1CCN(Cc2ccc(cc2)C(=O)Nc2ccc(C)c(Nc3nccc(n3)-c3cccnc3)c2)CC1"</span><span class="p">)</span>
<span class="n">Draw</span><span class="o">.</span><span class="n">MolToImage</span><span class="p">(</span><span class="n">smiles</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_18_0.png" src="../_images/talktorials_T010_binding_site_comparison_18_0.png"/>
</div>
</div>
<p>In order to inspect the 3D structure of STI, we use the open source tool <code class="docutils literal notranslate"><span class="pre">nglview</span></code>. Before we can view STI in <code class="docutils literal notranslate"><span class="pre">nglview</span></code>, we need to compute its 3D coordinates.</p>
<p>First, we add hydrogen atoms to the molecule, which are not always explicitly denoted in the SMILES format. Second, we use the distance geometry to obtain initial coordinates for the molecule and optimize the structure of the molecule using the force field UFF (Universal Force Field).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">molecule</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AddHs</span><span class="p">(</span><span class="n">smiles</span><span class="p">)</span>
<span class="n">molecule</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_20_0.png" src="../_images/talktorials_T010_binding_site_comparison_20_0.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">AllChem</span><span class="o">.</span><span class="n">EmbedMolecule</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
<span class="n">AllChem</span><span class="o">.</span><span class="n">UFFOptimizeMolecule</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
<span class="n">molecule</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_21_0.png" src="../_images/talktorials_T010_binding_site_comparison_21_0.png"/>
</div>
</div>
<p>Now, we are ready to roll in <code class="docutils literal notranslate"><span class="pre">nglview</span></code>!</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">show_rdkit</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "7ee4e44a50294c31b9b4b2d03d6f1ce1", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_25_0.png" src="../_images/talktorials_T010_binding_site_comparison_25_0.png"/>
</div>
</div>


<h3 id="Get-all-protein-STI-complexes-from-the-PDB">Get all protein-STI complexes from the PDB<a class="headerlink" href="#Get-all-protein-STI-complexes-from-the-PDB" title="Permalink to this headline">¬∂</a></h3>

<h4 id="Query-the-PDB">Query the PDB<a class="headerlink" href="#Query-the-PDB" title="Permalink to this headline">¬∂</a></h4>
<p>We can look up Imatinib/STI in open databases like the PDB and search for proteins that are reported targets. In the PDB, Imatinib is usually abbreviated to STI. We will search for both terms and merge the results in the following. We will use the <code class="docutils literal notranslate"><span class="pre">pypdb</span></code> package to perform the PDB queries.</p>
<p>First, we retrieve all proteins from the PDB that bind the ligand of interest.</p>
<ol class="arabic simple">
<li><p>Query PDB complexes by a ligand‚Äôs PDB HET code, here ‚ÄúSTI‚Äù.</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Set up query dictionary</span>
<span class="n">search_dict</span> <span class="o">=</span> <span class="n">pypdb</span><span class="o">.</span><span class="n">make_query</span><span class="p">(</span><span class="s2">"STI"</span><span class="p">)</span>
<span class="n">search_dict</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{'orgPdbQuery': {'queryType': 'AdvancedKeywordQuery',
  'description': 'Text Search for: STI',
  'keywords': 'STI'}}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Send query</span>
<span class="n">pdb_ids_by_ligand_code</span> <span class="o">=</span> <span class="n">pypdb</span><span class="o">.</span><span class="n">do_search</span><span class="p">(</span><span class="n">search_dict</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures connected with STI in the PDB: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">pdb_ids_by_ligand_code</span><span class="p">)</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="n">pdb_ids_by_ligand_code</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures connected with STI in the PDB: 64
1AVU 1AVW 1AVX 1BA7 1FPU 1IEP 1M52 1OPJ 1R8N 1R8O 1T45 1T46 1XBB 2BEA 2BEB 2DRE 2ESU 2ET2 2HYY 2OIQ 2PL0 3EA7 3EA8 3FW1 3GVU 3HEC 3I2A 3I2X 3K5V 3M3V 3MS9 3MSS 3OEZ 3PYY 3QQM 3S8J 3S8K 4BKJ 4CSV 4GCN 4GCO 4H9W 4HA2 4R7I 4TLP 5FNW 5FNX 5FZU 5FZY 5FZZ 5G00 5MQT 6HD4 6HD6 6I0I 6JOL 6KTN 6NPE 6NPU 6NPV 6OTU 6VXH 6WE5 6XRS
</pre></div></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Query PDB complexes by a ligand‚Äôs name, here ‚ÄúImatinib‚Äù.</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Set up query dictionary and send query</span>
<span class="n">search_dict</span> <span class="o">=</span> <span class="n">pypdb</span><span class="o">.</span><span class="n">make_query</span><span class="p">(</span><span class="s2">"Imatinib"</span><span class="p">)</span>
<span class="n">pdb_ids_by_ligand_name</span> <span class="o">=</span> <span class="n">pypdb</span><span class="o">.</span><span class="n">do_search</span><span class="p">(</span><span class="n">search_dict</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures connected with Imatinib in the PDB: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">pdb_ids_by_ligand_name</span><span class="p">)</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="n">pdb_ids_by_ligand_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures connected with Imatinib in the PDB: 42
1IEP 1M52 1OPJ 1T46 1XBB 2F4J 2GQG 2HYY 2OIQ 2PL0 2XYN 3EL7 3EL8 3FW1 3G0E 3G0F 3G6G 3G6H 3GVU 3HEC 3HEG 3K5V 3MS9 3MSS 3OEZ 3OF0 3PYY 3QLF 3QLG 4BKJ 4CSV 4R7I 5MQT 6HD4 6HD6 6JOL 6KTM 6KTN 6NPE 6NPU 6NPV 6VXH
</pre></div></div>
</div>
<p>Note that the query results can differ depending on the term used for the query ligand. We merge both query results and keep only unique entries.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pdb_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pdb_ids_by_ligand_code</span> <span class="o">+</span> <span class="n">pdb_ids_by_ligand_name</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures connected with STI/Imatinib in the PDB: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">pdb_ids</span><span class="p">)</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures connected with STI/Imatinib in the PDB: 78
</pre></div></div>
</div>


<h4 id="Filter-the-PDB-data-set">Filter the PDB data set<a class="headerlink" href="#Filter-the-PDB-data-set" title="Permalink to this headline">¬∂</a></h4>
<p>First, we retrieve meta information on the PDB structures.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">entity_info</span> <span class="o">=</span> <span class="p">[</span><span class="n">pypdb</span><span class="o">.</span><span class="n">get_entity_info</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">pdb_id</span> <span class="ow">in</span> <span class="n">pdb_ids</span><span class="p">]</span>

<span class="c1"># Show the first entry</span>
<span class="n">entity_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{'Method': {'@name': 'xray'},
 'Entity': {'@id': '1', '@type': 'protein', 'Chain': {'@id': 'A'}},
 'structureId': '3EL7',
 'bioAssemblies': '1',
 'release_date': 'Tue Oct 28 00:00:00 PDT 2008',
 'resolution': '2.80'}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Transform list to DataFrame</span>
<span class="n">entity_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">entity_info</span><span class="p">)</span>

<span class="c1"># Show the 5 first rows</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="n">entity_info</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures: 78
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>Method</th>
<th>Entity</th>
<th>structureId</th>
<th>bioAssemblies</th>
<th>release_date</th>
<th>resolution</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>{'@name': 'xray'}</td>
<td>{'@id': '1', '@type': 'protein', 'Chain': {'@i...</td>
<td>3EL7</td>
<td>1</td>
<td>Tue Oct 28 00:00:00 PDT 2008</td>
<td>2.80</td>
</tr>
<tr>
<th>1</th>
<td>{'@name': 'xray'}</td>
<td>{'@id': '1', '@type': 'protein', 'Chain': [{'@...</td>
<td>6XRS</td>
<td>2</td>
<td>Wed Sep 23 00:00:00 PDT 2020</td>
<td>2.80</td>
</tr>
<tr>
<th>2</th>
<td>{'@name': 'xray'}</td>
<td>{'@id': '1', '@type': 'protein', 'Chain': [{'@...</td>
<td>2BEA</td>
<td>2</td>
<td>Tue Jun 13 00:00:00 PDT 2006</td>
<td>2.35</td>
</tr>
<tr>
<th>3</th>
<td>{'@name': 'xray'}</td>
<td>{'@id': '1', '@type': 'protein', 'Chain': [{'@...</td>
<td>3QQM</td>
<td>8</td>
<td>Wed Mar 02 00:00:00 PST 2011</td>
<td>2.30</td>
</tr>
<tr>
<th>4</th>
<td>{'@name': 'xray'}</td>
<td>[{'@id': '1', '@type': 'protein', 'Chain': {'@...</td>
<td>1R8O</td>
<td>1</td>
<td>Tue May 25 00:00:00 PDT 2004</td>
<td>1.83</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>Now, we filter structures by the following criteria.</p>
<ol class="arabic simple">
<li><p>Filter by experimental method (‚Äúxray‚Äù).</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="n">entity_info</span><span class="p">[</span><span class="s2">"Method"</span><span class="p">]</span> <span class="o">==</span> <span class="p">{</span><span class="s2">"@name"</span><span class="p">:</span> <span class="s2">"xray"</span><span class="p">}]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures left: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures left: 77
</pre></div></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Filter by resolution (equal or lower than 3 ‚Ñ´).</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="n">entity_info</span><span class="p">[</span><span class="s2">"resolution"</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">3.0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures left: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures left: 76
</pre></div></div>
</div>
<ol class="arabic simple" start="3">
<li><p>Retain only PDB structures with a single chain (for simplicity).</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Remove all PDB IDs with multiple entities -- use pandas' apply function</span>
<span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="n">entity_info</span><span class="p">[</span><span class="s2">"Entity"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">dict</span><span class="p">))]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures left: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>

<span class="c1"># Remove all PDB IDs with multiple chains</span>
<span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="n">entity_info</span><span class="p">[</span><span class="s2">"Entity"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">"Chain"</span><span class="p">],</span> <span class="nb">dict</span><span class="p">))]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures left: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures left: 71
Number of structures left: 28
</pre></div></div>
</div>
<ol class="arabic simple" start="4">
<li><p>Retain only Imatinib-bound structures (e.g.¬†some PDB structures are returned that are associated with Imatinib but not bound to it).</p></li>
</ol>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nd">@redo</span><span class="o">.</span><span class="n">retriable</span><span class="p">(</span><span class="n">attempts</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">sleeptime</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">get_ligands</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">):</span>
    <span class="sd">"""Decorate pypdb.get_ligands so it retries after a failure"""</span>
    <span class="k">return</span> <span class="n">pypdb</span><span class="o">.</span><span class="n">get_ligands</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Add ligand information to DataFrame</span>
<span class="n">entity_info</span><span class="p">[</span><span class="s2">"ligandInfo"</span><span class="p">]</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="s2">"structureId"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">get_ligands</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="s2">"ligandInfo"</span><span class="p">]</span>
<span class="p">)</span>

<span class="c1"># Keep only PDB IDs with ligand information -- "~" negates boolean values</span>
<span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="o">~</span><span class="n">entity_info</span><span class="p">[</span><span class="s2">"ligandInfo"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)]</span>

<span class="c1"># Ligand information is given as dict (if only one ligand) and list, transform everything to list</span>
<span class="n">entity_info</span><span class="p">[</span><span class="s2">"ligandInfo"</span><span class="p">]</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="s2">"ligandInfo"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">"ligand"</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">"ligand"</span><span class="p">],</span> <span class="nb">list</span><span class="p">)</span> <span class="k">else</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s2">"ligand"</span><span class="p">]]</span>
<span class="p">)</span>

<span class="c1"># Keep only PDB IDs with STI ligand</span>
<span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span>
    <span class="n">entity_info</span><span class="p">[</span><span class="s2">"ligandInfo"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">"STI"</span> <span class="ow">in</span> <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="s2">"@chemicalID"</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
<span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures left: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>

<span class="c1"># After manual visual inspection remove 3GVU (contains 2x STI: not feasible for the automated workflow below)</span>
<span class="n">entity_info</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="n">entity_info</span><span class="p">[</span><span class="s2">"structureId"</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">"3GVU"</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Number of structures left: </span><span class="si">{</span><span class="n">entity_info</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of structures left: 9
Number of structures left: 8
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pdb_ids</span> <span class="o">=</span> <span class="n">entity_info</span><span class="p">[</span><span class="s2">"structureId"</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="s2">"Selected PDB IDs:"</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="n">pdb_ids</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Selected PDB IDs:
3HEC 4R7I 3FW1 6JOL 1T46 2PL0 1XBB 4CSV
</pre></div></div>
</div>
<p><strong>Note</strong>: The next step has technical reasons only. In order to maintain the notebooks automatically, we define here a certain set of PDB IDs to work with. Thus, even if new PDB structures matching our filtering criteria are added to the PDB, they will not be included in the downstream analysis and therefore the output will stay the same. Remote this step if you want to work with the full set of PDB structures.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pdb_ids</span> <span class="o">=</span> <span class="n">FROZEN_PDB_IDS</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">"Final set of PDB IDs:"</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="n">pdb_ids</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Final set of PDB IDs:
3HEC 2PL0 4CSV 4R7I 1XBB 3FW1 1T46
</pre></div></div>
</div>



<h3 id="Visualize-the-PDB-structures">Visualize the PDB structures<a class="headerlink" href="#Visualize-the-PDB-structures" title="Permalink to this headline">¬∂</a></h3>
<p>First, we load all structures in <code class="docutils literal notranslate"><span class="pre">nglview</span></code> for visual inspection of the 3D structure of the protein data set.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">NGLWidget</span><span class="p">()</span>
<span class="k">for</span> <span class="n">pdb_id</span> <span class="ow">in</span> <span class="n">pdb_ids</span><span class="p">:</span>
    <span class="n">view</span><span class="o">.</span><span class="n">add_pdbid</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c650008f1ed9410b9db65b78d57b0600", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_54_0.png" src="../_images/talktorials_T010_binding_site_comparison_54_0.png"/>
</div>
</div>
<p>Though this image is beautifully colorful and curly, it is not informative yet. We align the structures to each other in the next step.</p>


<h3 id="Align-the-PDB-structures-(full-protein)">Align the PDB structures (full protein)<a class="headerlink" href="#Align-the-PDB-structures-(full-protein)" title="Permalink to this headline">¬∂</a></h3>
<p>We will use one of our <code class="docutils literal notranslate"><span class="pre">opencadd</span></code> packages (<code class="docutils literal notranslate"><span class="pre">structure.superposition</span></code> module) to guide the structural alignment of the different proteins. The approach we will use is based on superposition guided by sequence alignment provided matched residues. There are other methods in the package, but this simple one will be enough to showcase some similarities.</p>
<p>Note: This approach biases the analysis towards structures with similar sequences. For an automated workflow (where we do not know the sequence or structural similarity of protein pairs) a solution could be to calculate the RMSD based on all three measures and retain the best for further analysis.</p>
<p>First, we show the alignment of all structures to the first structure in the our list.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Download PDB</span>
<span class="n">structures</span> <span class="o">=</span> <span class="p">[</span><span class="n">Structure</span><span class="o">.</span><span class="n">from_pdbid</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">pdb_id</span> <span class="ow">in</span> <span class="n">pdb_ids</span><span class="p">]</span>
<span class="c1"># Strip solvent and other artifacts of crystallography</span>
<span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">Structure</span><span class="o">.</span><span class="n">from_atomgroup</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">"protein"</span><span class="p">))</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">structures</span><span class="p">]</span>
<span class="c1"># Align proteins</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">align</span><span class="p">(</span><span class="n">proteins</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">METHODS</span><span class="p">[</span><span class="s2">"mda"</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">NGLWidget</span><span class="p">()</span>
<span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">proteins</span><span class="p">:</span>
    <span class="n">view</span><span class="o">.</span><span class="n">add_component</span><span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "88f3d66814db4909bc6d8a3e622f2f71", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[69]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_60_0.png" src="../_images/talktorials_T010_binding_site_comparison_60_0.png"/>
</div>
</div>
<p>The structural alignment for many helices is high, whereas lower or poor for other protein parts.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">calc_rmsd</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate RMSD between two structures.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    A : opencadd.structure.core.Structure</span>
<span class="sd">        Structure A.</span>
<span class="sd">    B : opencadd.structure.core.Structure</span>
<span class="sd">        Structure B.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        RMSD value.</span>
<span class="sd">    """</span>
    <span class="n">aligner</span> <span class="o">=</span> <span class="n">MDAnalysisAligner</span><span class="p">()</span>
    <span class="n">selection</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">matching_selection</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">)</span>
    <span class="n">A</span> <span class="o">=</span> <span class="n">A</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="n">selection</span><span class="p">[</span><span class="s2">"reference"</span><span class="p">])</span>
    <span class="n">B</span> <span class="o">=</span> <span class="n">B</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="n">selection</span><span class="p">[</span><span class="s2">"mobile"</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">rms</span><span class="o">.</span><span class="n">rmsd</span><span class="p">(</span><span class="n">A</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">B</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">superposition</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">calc_rmsd_matrix</span><span class="p">(</span><span class="n">structures</span><span class="p">,</span> <span class="n">names</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate RMSD matrix between a list of structures.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    structures : list of opencadd.structure.core.Structure</span>
<span class="sd">        List of structures.</span>
<span class="sd">    names : list of str</span>
<span class="sd">        List of structure names.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pandas.DataFrame</span>
<span class="sd">        RMSD matrix.</span>
<span class="sd">    """</span>
    <span class="n">values</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="p">{}</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">name_i</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">structures</span><span class="p">,</span> <span class="n">names</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="p">(</span><span class="n">B</span><span class="p">,</span> <span class="n">name_j</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">structures</span><span class="p">,</span> <span class="n">names</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">j</span><span class="p">:</span>
                <span class="n">values</span><span class="p">[</span><span class="n">name_i</span><span class="p">][</span><span class="n">name_j</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="n">j</span><span class="p">:</span>
                <span class="n">rmsd</span> <span class="o">=</span> <span class="n">calc_rmsd</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">)</span>
                <span class="n">values</span><span class="p">[</span><span class="n">name_i</span><span class="p">][</span><span class="n">name_j</span><span class="p">]</span> <span class="o">=</span> <span class="n">rmsd</span>
                <span class="n">values</span><span class="p">[</span><span class="n">name_j</span><span class="p">][</span><span class="n">name_i</span><span class="p">]</span> <span class="o">=</span> <span class="n">rmsd</span>
                <span class="k">continue</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</pre></div>
</div>
</div>


<h3 id="Get-pairwise-RMSD-(full-protein)">Get pairwise RMSD (full protein)<a class="headerlink" href="#Get-pairwise-RMSD-(full-protein)" title="Permalink to this headline">¬∂</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_CHECK_OUTPUT</span>
<span class="n">rmsd_matrix</span> <span class="o">=</span> <span class="n">calc_rmsd_matrix</span><span class="p">(</span><span class="n">proteins</span><span class="p">,</span> <span class="n">pdb_ids</span><span class="p">)</span>
<span class="n">rmsd_matrix</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>3HEC</th>
<th>2PL0</th>
<th>4CSV</th>
<th>4R7I</th>
<th>1XBB</th>
<th>3FW1</th>
<th>1T46</th>
</tr>
</thead>
<tbody>
<tr>
<th>3HEC</th>
<td>0.0</td>
<td>16.1</td>
<td>12.6</td>
<td>20.5</td>
<td>17.2</td>
<td>23.7</td>
<td>21.7</td>
</tr>
<tr>
<th>2PL0</th>
<td>16.1</td>
<td>0.0</td>
<td>3.7</td>
<td>13.5</td>
<td>7.5</td>
<td>22.8</td>
<td>14.1</td>
</tr>
<tr>
<th>4CSV</th>
<td>12.6</td>
<td>3.7</td>
<td>0.0</td>
<td>14.0</td>
<td>7.8</td>
<td>22.5</td>
<td>14.6</td>
</tr>
<tr>
<th>4R7I</th>
<td>20.5</td>
<td>13.5</td>
<td>14.0</td>
<td>0.0</td>
<td>11.9</td>
<td>21.9</td>
<td>3.2</td>
</tr>
<tr>
<th>1XBB</th>
<td>17.2</td>
<td>7.5</td>
<td>7.8</td>
<td>11.9</td>
<td>0.0</td>
<td>22.2</td>
<td>12.6</td>
</tr>
<tr>
<th>3FW1</th>
<td>23.7</td>
<td>22.8</td>
<td>22.5</td>
<td>21.9</td>
<td>22.2</td>
<td>0.0</td>
<td>21.6</td>
</tr>
<tr>
<th>1T46</th>
<td>21.7</td>
<td>14.1</td>
<td>14.6</td>
<td>3.2</td>
<td>12.6</td>
<td>21.6</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>We visualize the results of this RMSD refinement as heatmap.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Make sure matplotlib version &gt;= 3.1.2; otherwise you'll get Y-cropped heatmaps</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="n">rmsd_matrix</span><span class="p">,</span>
    <span class="n">linewidths</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">"label"</span><span class="p">:</span> <span class="s2">"RMSD ($\AA$)"</span><span class="p">},</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s2">"Blues"</span><span class="p">,</span>
<span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_67_0.png" src="../_images/talktorials_T010_binding_site_comparison_67_0.png"/>
</div>
</div>
<p>We cluster the heatmap in order to see protein similarity based on the RMSD refinement.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">plot_clustermap</span><span class="p">(</span><span class="n">rmsd</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Plot clustered heatmap from import RMSD matrix.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    rmsd : pandas.DataFrame</span>
<span class="sd">        RMSD matrix.</span>
<span class="sd">    title : str</span>
<span class="sd">        Plot title.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.figure.Figure</span>
<span class="sd">        Clustered heatmap.</span>
<span class="sd">    """</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span>
        <span class="n">rmsd</span><span class="p">,</span>
        <span class="n">linewidths</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">"label"</span><span class="p">:</span> <span class="s2">"RMSD ($\AA$)"</span><span class="p">},</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s2">"Blues"</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_clustermap</span><span class="p">(</span><span class="n">rmsd_matrix</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_70_0.png" src="../_images/talktorials_T010_binding_site_comparison_70_0.png"/>
</div>
</div>
<p>The RMSD comparison shows that one protein differs most from the other proteins, i.e.¬†3FW1. Let‚Äôs try to understand by checking which proteins we have in our dataset.</p>
<p>Proteins are classified by the chemical reactions that they catalyze with EC (Enzyme Commission) numbers. We will use them here to check the enzymatic groups the proteins belong to.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_CHECK_OUTPUT</span>
<span class="c1"># Get EC numbers for PDB IDs from PDB</span>
<span class="n">pdbs_info</span> <span class="o">=</span> <span class="p">[</span><span class="n">pypdb</span><span class="o">.</span><span class="n">get_all_info</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">)</span> <span class="k">for</span> <span class="n">pdb_id</span> <span class="ow">in</span> <span class="n">pdb_ids</span><span class="p">]</span>
<span class="n">ec_numbers</span> <span class="o">=</span> <span class="p">[</span><span class="n">pdb_info</span><span class="p">[</span><span class="s2">"polymer"</span><span class="p">][</span><span class="s2">"enzClass"</span><span class="p">][</span><span class="s2">"@ec"</span><span class="p">]</span> <span class="k">for</span> <span class="n">pdb_info</span> <span class="ow">in</span> <span class="n">pdbs_info</span><span class="p">]</span>
<span class="n">ec_numbers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">"pdb_id"</span><span class="p">:</span> <span class="n">pdb_ids</span><span class="p">,</span> <span class="s2">"ec_number"</span><span class="p">:</span> <span class="n">ec_numbers</span><span class="p">}</span>
<span class="n">ec_numbers</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ec_numbers</span><span class="p">)</span>
<span class="n">ec_numbers</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>pdb_id</th>
<th>ec_number</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>3HEC</td>
<td>2.7.11.24</td>
</tr>
<tr>
<th>1</th>
<td>2PL0</td>
<td>2.7.10.2</td>
</tr>
<tr>
<th>2</th>
<td>4CSV</td>
<td>2.7.10.2</td>
</tr>
<tr>
<th>3</th>
<td>4R7I</td>
<td>2.7.10.1</td>
</tr>
<tr>
<th>4</th>
<td>1XBB</td>
<td>2.7.10.2</td>
</tr>
<tr>
<th>5</th>
<td>3FW1</td>
<td>1.10.5.1</td>
</tr>
<tr>
<th>6</th>
<td>1T46</td>
<td>2.7.10.1</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>We can see that 3FW1, the human quinone reductase 2 (NQO2), belongs to EC class 1, i.e.¬†oxidoreductases, whereas the other proteins belong to class 2.7, i.e.¬†phosphorus transferases, which contain the tyrosine kinases (EC 2.7.10.2), the designated targets for Imatinib. 3FW1 is a reported off-target ‚Äúwith potential implications for drug design and treatment of chronic myelogenous leukemia in patients‚Äù (<a class="reference external" href="https://bmcstructbiol.biomedcentral.com/articles/10.1186/1472-6807-9-7">BMC Struct. Biol. (2009),
9</a>).</p>


<h3 id="Align-PDB-structures-(binding-sites)">Align PDB structures (binding sites)<a class="headerlink" href="#Align-PDB-structures-(binding-sites)" title="Permalink to this headline">¬∂</a></h3>
<p>So far, we have used the full protein structure for the alignment and RMSD refinement. However, ligands bind only at a protein‚Äôs binding site. Therefore, the similarity of binding sites is a more putative basis for off-target prediction rather than of the similarity of full protein structures.</p>
<p>We define a binding site of a protein by selecting all residues that are within 10 ‚Ñ´ of any ligand atom. These binding site residues are used for alignment and only their C…ë atoms (protein backbone) are used for the RMSD refinement. Here, we show the alignment of all structures to the first structure in our list.</p>
<p>Note that for the aligner to work, full residues must be selected as part of the distance search, hence the need for the <code class="docutils literal notranslate"><span class="pre">same</span> <span class="pre">residue</span> <span class="pre">as</span> <span class="pre">(...)</span></code> query.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sel</span> <span class="o">=</span> <span class="s2">"same residue as (resname STI or (around 10 resname STI))"</span>
<span class="n">binding_sites</span> <span class="o">=</span> <span class="p">[</span><span class="n">Structure</span><span class="o">.</span><span class="n">from_atomgroup</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="n">sel</span><span class="p">))</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">structures</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>Let‚Äôs look at the proteins‚Äô binding sites only.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">NGLWidget</span><span class="p">()</span>
<span class="k">for</span> <span class="n">binding_site</span> <span class="ow">in</span> <span class="n">binding_sites</span><span class="p">:</span>
    <span class="n">view</span><span class="o">.</span><span class="n">add_component</span><span class="p">(</span><span class="n">binding_site</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "686a7ac63f21452c8ff701a460115d46", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_81_0.png" src="../_images/talktorials_T010_binding_site_comparison_81_0.png"/>
</div>
</div>
<p>Now we align the binding sites.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">results_binding_sites</span> <span class="o">=</span> <span class="n">align</span><span class="p">(</span><span class="n">binding_sites</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">METHODS</span><span class="p">[</span><span class="s2">"mda"</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>Let‚Äôs visualize our aligned binding sites.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">NGLWidget</span><span class="p">()</span>
<span class="k">for</span> <span class="n">binding_site</span> <span class="ow">in</span> <span class="n">binding_sites</span><span class="p">:</span>
    <span class="n">view</span><span class="o">.</span><span class="n">add_component</span><span class="p">(</span><span class="n">binding_site</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d9ae3fc36bb245cda4b67fdafcc83a6f", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[75]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[76]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[76]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_87_0.png" src="../_images/talktorials_T010_binding_site_comparison_87_0.png"/>
</div>
</div>


<h3 id="Get-pairwise-RMSD-(binding-sites)">Get pairwise RMSD (binding sites)<a class="headerlink" href="#Get-pairwise-RMSD-(binding-sites)" title="Permalink to this headline">¬∂</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_CHECK_OUTPUT</span>
<span class="n">rmsd_matrix_binding_sites</span> <span class="o">=</span> <span class="n">calc_rmsd_matrix</span><span class="p">(</span><span class="n">binding_sites</span><span class="p">,</span> <span class="n">pdb_ids</span><span class="p">)</span>
<span class="n">rmsd_matrix_binding_sites</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>3HEC</th>
<th>2PL0</th>
<th>4CSV</th>
<th>4R7I</th>
<th>1XBB</th>
<th>3FW1</th>
<th>1T46</th>
</tr>
</thead>
<tbody>
<tr>
<th>3HEC</th>
<td>0.0</td>
<td>6.2</td>
<td>4.6</td>
<td>5.7</td>
<td>12.3</td>
<td>15.0</td>
<td>9.8</td>
</tr>
<tr>
<th>2PL0</th>
<td>6.2</td>
<td>0.0</td>
<td>4.0</td>
<td>4.0</td>
<td>11.7</td>
<td>11.3</td>
<td>4.3</td>
</tr>
<tr>
<th>4CSV</th>
<td>4.6</td>
<td>4.0</td>
<td>0.0</td>
<td>5.2</td>
<td>10.7</td>
<td>13.9</td>
<td>5.9</td>
</tr>
<tr>
<th>4R7I</th>
<td>5.7</td>
<td>4.0</td>
<td>5.2</td>
<td>0.0</td>
<td>10.5</td>
<td>14.5</td>
<td>2.7</td>
</tr>
<tr>
<th>1XBB</th>
<td>12.3</td>
<td>11.7</td>
<td>10.7</td>
<td>10.5</td>
<td>0.0</td>
<td>13.5</td>
<td>11.1</td>
</tr>
<tr>
<th>3FW1</th>
<td>15.0</td>
<td>11.3</td>
<td>13.9</td>
<td>14.5</td>
<td>13.5</td>
<td>0.0</td>
<td>15.3</td>
</tr>
<tr>
<th>1T46</th>
<td>9.8</td>
<td>4.3</td>
<td>5.9</td>
<td>2.7</td>
<td>11.1</td>
<td>15.3</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>We show the clustered heatmap for the RMSD results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Show the pairwise RMSD values as clustered heatmap</span>
<span class="n">plot_clustermap</span><span class="p">(</span><span class="n">rmsd_matrix_binding_sites</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_91_0.png" src="../_images/talktorials_T010_binding_site_comparison_91_0.png"/>
</div>
</div>
<p>What are the key observations in this heatmap?</p>
<ul class="simple">
<li><p>As observed already during the <em>full protein</em> comparison, also the <em>binding site</em> comparison reveals the highest dissimilarity for <code class="docutils literal notranslate"><span class="pre">3FW1</span></code>. Since <code class="docutils literal notranslate"><span class="pre">3FW1</span></code> is the only structure in our dataset representing not a kinase, we are content that our similarity measure was able to spot this. As discussed before, the human quinone reductase 2 (<code class="docutils literal notranslate"><span class="pre">3FW1</span></code>) is a reported off-target for Imatinib.</p></li>
<li><p>Also the kinase <code class="docutils literal notranslate"><span class="pre">1XBB</span></code> shows a higher dissimilarity compared to the other kinases. This can be explained by <code class="docutils literal notranslate"><span class="pre">1XBB</span></code> being resolved in a different conformation (DFG-in) than the other kinases (DFG-out). The DFG motif is an important structural element in kinases, defining whether a kinase is active (DFG-in conformation) or inactive (DFG-out conformation). Note: DFG conformations of kinase structures can be looked up e.g.¬†in the <a class="reference external" href="https://www.klifs.net">KLIFS database</a>.</p></li>
<li><p>The remaining structures are comparatively similar to each other, which we would expect since they all represent kinases in the overall same conformation.</p></li>
</ul>
<p>Note that RMSD values as calculated here are dependent on the residue selection (binding site definition) and the quality of the a priori sequence alignment.</p>


<h3 id="Filter-out-outliers">Filter out outliers<a class="headerlink" href="#Filter-out-outliers" title="Permalink to this headline">¬∂</a></h3>
<p>For a cleaner depiction of the most similar structures, we can filter <code class="docutils literal notranslate"><span class="pre">3FW1</span></code> and <code class="docutils literal notranslate"><span class="pre">1XBB</span></code> out and re-run the binding site comparison.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">filtered_structures</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">filtered_pdb_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">structure</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pdb_ids</span><span class="p">,</span> <span class="n">structures</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">"3FW1"</span><span class="p">,</span> <span class="s2">"1XBB"</span><span class="p">):</span>
        <span class="n">filtered_structures</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>
        <span class="n">filtered_pdb_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selection</span> <span class="o">=</span> <span class="s2">"same residue as (resname STI or (around 10 resname STI))"</span>
<span class="n">filtered_binding_sites</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">Structure</span><span class="o">.</span><span class="n">from_atomgroup</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="n">selection</span><span class="p">))</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">filtered_structures</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[77]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">NGLWidget</span><span class="p">()</span>
<span class="k">for</span> <span class="n">binding_site</span> <span class="ow">in</span> <span class="n">filtered_binding_sites</span><span class="p">:</span>
    <span class="n">view</span><span class="o">.</span><span class="n">add_component</span><span class="p">(</span><span class="n">binding_site</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "f7680f784b134c4ea831b3844208cd90", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[78]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[79]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[79]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_98_0.png" src="../_images/talktorials_T010_binding_site_comparison_98_0.png"/>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">filtered_results_binding_sites</span> <span class="o">=</span> <span class="n">align</span><span class="p">(</span><span class="n">filtered_binding_sites</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">METHODS</span><span class="p">[</span><span class="s2">"mda"</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[80]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nv</span><span class="o">.</span><span class="n">NGLWidget</span><span class="p">()</span>
<span class="k">for</span> <span class="n">binding_site</span> <span class="ow">in</span> <span class="n">filtered_binding_sites</span><span class="p">:</span>
    <span class="n">view</span><span class="o">.</span><span class="n">add_component</span><span class="p">(</span><span class="n">binding_site</span><span class="o">.</span><span class="n">atoms</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d7eaa13c40954bbebd62128a4df256a4", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[81]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">render_image</span><span class="p">(</span><span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[82]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">view</span><span class="o">.</span><span class="n">_display_image</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[82]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_102_0.png" src="../_images/talktorials_T010_binding_site_comparison_102_0.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_CHECK_OUTPUT</span>
<span class="n">filtered_rmsd_matrix_bs</span> <span class="o">=</span> <span class="n">calc_rmsd_matrix</span><span class="p">(</span><span class="n">filtered_binding_sites</span><span class="p">,</span> <span class="n">filtered_pdb_ids</span><span class="p">)</span>
<span class="n">filtered_rmsd_matrix_bs</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>3HEC</th>
<th>2PL0</th>
<th>4CSV</th>
<th>4R7I</th>
<th>1T46</th>
</tr>
</thead>
<tbody>
<tr>
<th>3HEC</th>
<td>0.0</td>
<td>6.2</td>
<td>4.6</td>
<td>5.7</td>
<td>9.8</td>
</tr>
<tr>
<th>2PL0</th>
<td>6.2</td>
<td>0.0</td>
<td>4.0</td>
<td>4.0</td>
<td>4.3</td>
</tr>
<tr>
<th>4CSV</th>
<td>4.6</td>
<td>4.0</td>
<td>0.0</td>
<td>5.2</td>
<td>5.9</td>
</tr>
<tr>
<th>4R7I</th>
<td>5.7</td>
<td>4.0</td>
<td>5.2</td>
<td>0.0</td>
<td>2.7</td>
</tr>
<tr>
<th>1T46</th>
<td>9.8</td>
<td>4.3</td>
<td>5.9</td>
<td>2.7</td>
<td>0.0</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_clustermap</span><span class="p">(</span><span class="n">filtered_rmsd_matrix_bs</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T010_binding_site_comparison_104_0.png" src="../_images/talktorials_T010_binding_site_comparison_104_0.png"/>
</div>
</div>
<p>Much better!</p>



<h2 id="Discussion">Discussion<a class="headerlink" href="#Discussion" title="Permalink to this headline">¬∂</a></h2>
<p>In this talktorial, we have used a simple comparison approach, i.e.¬†sequence alignment and subsequent RMSD refinement of (i) full proteins and (ii) binding sites, to assess the similarity and dissimilarity of a small set of structures showing Imatinib-binding proteins. In our data set, we were able to spot an off-target for Imatinib (highest dissimilarity) and a kinase resolved in a different conformation compared to the rest of the kinases (also with a relatively high dissimilarity). Given our
simple approach, we are content with these results!</p>
<p>In a real case scenario, off-targets for Imatinib would be predicted by comparing the binding site of an intended target of Imatinib (a tyrosine kinase) with a large database of resolved structures (PDB). Since this results in the comparison of sequences also with low similarity, more sophisticated methods should be invoked that use a sequence-independent alignment algorithm and that include the physico-chemical properties of the binding site.</p>


<h2 id="Quiz">Quiz<a class="headerlink" href="#Quiz" title="Permalink to this headline">¬∂</a></h2>
<ol class="arabic simple">
<li><p>Explain the terms on- and off-targets of a drug.</p></li>
<li><p>Explain why binding site similarity can be used to find off-targets based on a query target.</p></li>
<li><p>Discuss how useful the RMSD value of (i) full proteins and (ii) protein binding sites is for off-target prediction.</p></li>
<li><p>Think of alternate approaches to encode binding site information.</p></li>
</ol>




          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
            <a href="T009_compound_ensemble_pharmacophores.html" title="T009 ¬∑ Ligand-based pharmacophores"
               class="md-flex md-footer-nav__link md-footer-nav__link--prev"
               rel="prev">
              <div class="md-flex__cell md-flex__cell--shrink">
                <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
              </div>
              <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
                <span class="md-flex__ellipsis">
                  <span
                      class="md-footer-nav__direction"> Previous </span> T009 ¬∑ Ligand-based pharmacophores </span>
              </div>
            </a>
          
          
            <a href="../all_talktorials.html" title="Complete list of talktorials"
               class="md-flex md-footer-nav__link md-footer-nav__link--next"
               rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span
                class="md-flex__ellipsis"> <span
                class="md-footer-nav__direction"> Next </span> Complete list of talktorials </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink"><i
                class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2018-2021, Volkamer Lab. Project structure based on the Computational Molecular Science Python Cookiecutter version 1.1.
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 3.5.1.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="../_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>