{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Talktorial 11 (part B)\n",
    "\n",
    "# Structure-based CADD using online APIs/servers\n",
    "\n",
    "__Developed at AG Volkamer, Charité__\n",
    "\n",
    "Dr. Jaime Rodríguez-Guerra, Dominique Sydow\n",
    "\n",
    "## Aim of this talktorial\n",
    "\n",
    "> This is part B of the \"Online webservices\" talktorial:\n",
    ">\n",
    "> - 11a. Querying KLIFS & PubChem for potential kinase inhibitors\n",
    "> - __11b. Docking the candidates against the target obtained in 11a__\n",
    "> - 11c. Assessing the results and comparing against known data\n",
    "\n",
    "\n",
    "After obtaining input structures we will use molecular docking software to find good protein-ligand poses.\n",
    "\n",
    "## Learning goals\n",
    "\n",
    "### Theory\n",
    "\n",
    "- Molecular docking basics\n",
    "- Available software\n",
    "- Known limitations\n",
    "\n",
    "### Practical\n",
    "\n",
    "- Prepare the structures\n",
    "- Guess the binding site\n",
    "- Run the docking calculation\n",
    "- Save the results\n",
    "\n",
    "## References\n",
    "\n",
    "* Chapter on Structure‐Based Virtual Screening, [Ch. 6.8 in \"Applied Chemoinformatics: Achievements and Future Opportunities\" (2018)](https://onlinelibrary.wiley.com/doi/book/10.1002/9783527806539)\n",
    "* How to benchmark docking software ([_J. Med. Chem._ (2006), __49__, 23, 6789–6801](https://pubs.acs.org/doi/abs/10.1021/jm0608356))\n",
    "* A review on molecular docking software ([_Biophysical Reviews_ (2017), __9__, 2, 91–102](https://www.ncbi.nlm.nih.gov/pubmed/28510083))\n",
    "* DoGSiteScorer, a program to identify binding sites.\n",
    "   * [_J. Chem. Inf. Model._ (2010), __50__, 11, 2041-2052](https://doi.org/10.1021/ci100241y)\n",
    "   * [_Bioinformatics_ (2012), __28__, 15, 2074–2075](https://doi.org/10.1093/bioinformatics/bts310)\n",
    "* SwissDock, a docking program.\n",
    "   * [_Nucleic Acids Res._ (2011), __39__, W270-7.](https://academic.oup.com/nar/article/39/suppl_2/W270/2506492)\n",
    "   * [_J Comput Chem._ (2011), __32__, 10, 2149-59](https://onlinelibrary.wiley.com/doi/abs/10.1002/jcc.21797)\n",
    "* OPAL webservices, which can run software remotely on demand\n",
    "   * Manuscript: [_Nucleic Acids Res._ (2010), __38__, W724-31](https://academic.oup.com/nar/article/38/suppl_2/W724/1122840)\n",
    "   * Documentation: [Opal: Simple Web Services Wrappers for Scientific Applications](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.533.7960&rep=rep1&type=pdf)\n",
    "* AutoDock Vina, a docking program ([_J Comput Chem._ (2010), __31__, 455–461](https://onlinelibrary.wiley.com/doi/abs/10.1002/jcc.21334))\n",
    "* Molecular Docking in Modern Drug Discovery: Principles and Recent Applications ([_Mallika Alvala's Lab_ (2019)](https://www.researchgate.net/publication/335022707_Molecular_Docking_in_Modern_Drug_Discovery_Principles_and_Recent_Applications))\n",
    "* Software for molecular docking: a review ([_Biophys Rev._ (2017), __9(2)__, 91-102](https://www.ncbi.nlm.nih.gov/pubmed/28510083))\n",
    "* Molecular Docking: A powerful approach for structure-based drug discovery ([Curr Comput Aided Drug Des._ (2011), __7(2)__, 146–157.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3151162/))\n",
    "* Progress in molecular docking ([_Fan, J., Fu, A. & Zhang, L. Quant Biol_ (2019), __7__, 83.](https://doi.org/10.1007/s40484-019-0172-y))\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Theory\n",
    "\n",
    "### Molecular docking\n",
    "\n",
    "Adapted from [Pagadala, Syed & Tuszynski, Biophysical Reviews 2017, 9, 2, 91–102](https://link.springer.com/article/10.1007%2Fs12551-016-0247-1):\n",
    "\n",
    "> Molecular docking methodology explores the behavior of small molecules in the binding site of a target protein. The applied software performs a search algorithm in which the conformation of the ligand is evaluated recursively until the convergence to the minimum energy is reached. Finally, an affinity scoring function, ΔG (U total in kcal/mol), is employed to rank the candidate poses as the sum of the electrostatic and van der Waals energies. The driving forces for these specific interactions in biological systems aim toward complementarities between the shape and electrostatics of the binding site surfaces and the ligand or substrate.\n",
    "\n",
    "These scoring functions are tailored to be fast to compute, and thus their accuracy is often poorer than in other molecular modeling methods. Normally, they are based on:\n",
    "\n",
    "- Molecular mechanics principles\n",
    "- Knowledge-based potentials\n",
    "- Shape and geometric complementarity\n",
    "\n",
    "To reduce the search space dimensionality, some approximations are often applied:\n",
    "\n",
    "- The protein structure is considered mostly rigid, with maybe some side chains in the vicinities of the search area allowed to explore a limited set of conformations (rotamers).\n",
    "- The ligand can be considered rigid in virtual screening studies, but it is often allowed to freely explore its bond torsions in more detailed calculations. A compromise between both options is defining a set of probable conformations beforehand.\n",
    "\n",
    "#### Examples of existing software\n",
    "\n",
    "- Commercial\n",
    "    - GOLD\n",
    "    - Schrödinger\n",
    "    - FlexX\n",
    "- Free (or free for academics)\n",
    "    - AutoDock\n",
    "    - AutoDock Vina\n",
    "    - DOCK\n",
    "    - OpenEye"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Known limitations\n",
    "\n",
    "The following approximations could introduce artifacts in the calculation. \n",
    "\n",
    "- Since the protein is mostly rigid, the dynamic, adaptive nature of the protein-ligand binding is barely explored. This can result in some false positives: even if the ligand finds a suitable pose in the binding pocket, this position is not guaranteed until the protein is allowed to explore near-minima conformations. In other words, short molecular dynamics trajectories are always recommended to check that the ligand stays in the binding pocket.\n",
    "- The scoring function must be cheap to resolve. While the accuracy is good enough to distinguish good poses from bad poses, it can have problems sorting the best poses. For example, while most popular docking programs are able to find the experimental pose in their calculations, this pose is rarely the best one of the proposed set.\n",
    "- To reduce the computational cost, docking procedures are only performed in a subset of the protein (normally around a known binding pocket). Choosing the correct binding pocket becomes then another question in the CADD pipeline. \n",
    "- To leverage the accuracy of the calculation, the structures must be reduced accordingly. Protonation states of amino acids and the ligands can be tricky to get right, especially in the case of (potential) tautomers. This introduces yet another cause to obtain wrong results.\n",
    "\n",
    "Despite all these limitations, docking calculations are still popularly used in all pharma laboratories, along with other types of molecular simulation. In this part of the talktorial we will learn how to:\n",
    "\n",
    "1. Prepare the protein and ligands obtained in part A  (locally)\n",
    "2. Estimate the most probable binding pocket (online)\n",
    "3. Run the docking calculations (online)\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Molecular docking\n",
    "\n",
    "Molecular docking is a key instrument in modern drug research. The aim is to model the interaction between a small molecule and a protein at the atomic level and thus to characterize the behavior of small molecules in the binding site of target proteins and to explain basic biochemical processes. The docking methods can be used to calculate the drugability of the compounds and their specificity to a specific target for further lead optimization processes. First, a sampling algorithm is performed which recursively evaluates the conformation of the ligands and their position and orientation within these sites until a minimum energy is reached. The binding affinities of the ligand conformations are then evaluated using a scoring function. \n",
    "\n",
    "<img src=\"images/docking.png\" align=\"above\" alt=\"Image cannot be shown\" width=\"650\">\n",
    "<div align=\"center\"> Figure 1: Molecular docking workflow</div>\n",
    "\n",
    "\n",
    "### Docking methodologies\n",
    "\n",
    "1. Rigid ligand and rigid receptor docking\n",
    "\n",
    "    - Initially, both the ligand and the receptor were considered rigid in molecular docking. However, the search space was very limited. The ligand flexibility was generated by using a precalculated set of ligand conformations or by considering a certain atom-atom overlap between protein and ligand.\n",
    "    \n",
    "\n",
    "2. Flexible ligand and flexible receptor docking\n",
    "\n",
    "    - Both ligand and receptor change their conformations to form an energetically minimal perfect fit complex. Therefore, it makes sense to consider ligands and receptors as flexible. However, the integration of receptor flexibility is a big challenge. To achieve this, Monte Carlo simulations are used. But this is associated with a very high computing effort. \n",
    "    \n",
    "\n",
    "3. Flexible ligand and rigid receptor docking\n",
    "\n",
    "    - Taking into account the flexibility of both the ligand and the receptor is very costly, though. Therefore, a compromise between accuracy and computing time is usually used, where the ligand is considered flexible and the receptor rigid during docking.\n",
    "\n",
    "### Sampling algorithms\n",
    "\n",
    "Almost all current tools use a combination of several of the following algorithms for their molecular docking method.\n",
    "\n",
    "1. Matching algorithms\n",
    "\n",
    "    - The matching algorithms (MA) find molecules that have a high degree of shape similarity to the pockets or binding sites. Programs based on MA have the advantage of speed. There are other limitations such as the prior need for detailed receptor geometry and a lack of molecular flexibility that does not precisely define many aspects of ligand-protein interactions.\n",
    "    \n",
    "\n",
    "2. Incremental construction\n",
    "\n",
    "    - In the incremental construction, the ligand is fragmented into different segments by breaking its rotatable bonds and incrementally brought into an active position. One of the segments is anchored to the receptor surface. The anchor is generally regarded as the fragment that shows maximum interactions with the receptor surface, has a minimal number of alternative conformations, and is fairly rigid. The algorithm is extremely fast and robust and can realize the flexibility of the ligand. A disadvantage, however, is that it is limited to medium-sized ligands, as the number of fragments generated is a problem with large ligands.\n",
    "    \n",
    "\n",
    "3. Stochastic methods search\n",
    "\n",
    "    - Stochastic methods search the conformational space by randomly modifying a ligand conformation or a population of ligands. Two examples are Monte Carlo methods and genetic algorithms.\n",
    "    \n",
    "    3.1 Monte Carlo methods\n",
    "    \n",
    "      - The Monte Carlo method involves modifying ligands through step-by-step bond rotation, rigid body displacement and rotation. The resulting conformity is tested with an energy-based selection criterion.  If it is fulfilled, this conformity is stored and further modified to generate the next conformation until a predefined number of result conformations is reached. \n",
    "      \n",
    "    3.2 Genetic algorithms\n",
    "    \n",
    "      - The genetic algorithm is mainly used to find the global minima and goes back to Darwin's theory of evolution. The degrees of freedom of the ligand are encoded as binary strings called genes. These in turn form the chromosome, which is actually the pose of the ligand. The ligand is then altered by mutation or crossover.\n",
    "    \n",
    "\n",
    "4. Molecular dynamics\n",
    "\n",
    "    - The molecular dynamic allows the flexibility of both the ligand and the protein to be represented, since each individual atom can be moved in the region of the rest atom. Since Monte Carlo simulations can only make small steps, the simulations are slow, but can be used very well for local optimization.\n",
    "    \n",
    "\n",
    "### Scoring functions\n",
    "\n",
    "Scoring functions are used to estimate the binding energy of a complex as accurately as possible and at the same time require little computing time. They contain estimates and are based on many assumptions and simplifications, but therefore require less computational costs and time than if an exact calculation of the binding affinity between the protein and the ligand were carried out.\n",
    "\n",
    "1. Force-field-based scoring functions\n",
    "\n",
    "    - Force-field-based scoring functions are based on physical atomic interactions (e.g. Van der Waals interactions). They evaluate the binding energy by calculating the sum of the unbound interactions.\n",
    "    \n",
    "\n",
    "2. Empirical scoring functions\n",
    "\n",
    "    - With empirical scoring functions, the binding energy is broken down in several energy components. Energy components are e.g. hydrogen bond, ion interaction, hydrophobic effect or bond entropy. The single components of a complex are multiplied by a coefficient and combined to a final result. The coefficients are obtained from a regression analysis.\n",
    "    \n",
    "\n",
    "3. Knowledge-based scoring functions\n",
    "\n",
    "    - Knowledge-based scoring functions uses statistical analysis of crystal structures of ligand-protein complexes. The idea is based on the assumption that the stronger the interaction, the greater the frequency of occurrence. Thus, the interatomic contact frequencies between the protein and the ligand can be derived from the structural information.\n",
    "\n",
    "\n",
    "### Known limitations\n",
    "\n",
    "The following approximations could introduce artifacts in the calculation. \n",
    "\n",
    "- Since the protein is mostly rigid, the dynamic, adaptive nature of the protein-ligand binding is barely explored. This can result in some false positives: even if the ligand finds a suitable pose in the binding pocket, this position is not guaranteed until the protein is allowed to explore near-minima conformations. In other words, short molecular dynamics trajectories are always recommended to check that the ligand stays in the binding pocket.\n",
    "- The scoring function must be cheap to resolve. While the accuracy is good enough to distinguish good poses from bad poses, it can have problems sorting the best poses. For example, while most popular docking programs are able to find the experimental pose in their calculations, this pose is rarely the best one of the proposed set.\n",
    "- To reduce the computational cost, docking procedures are only performed in a subset of the protein (normally around a known binding pocket). Choosing the correct binding pocket becomes then another question in the CADD pipeline. \n",
    "- To leverage the accuracy of the calculation, the structures must be reduced accordingly. Protonation states of amino acids and the ligands can be tricky to get right, especially in the case of (potential) tautomers. This introduces yet another cause to obtain wrong results.\n",
    "\n",
    "#### Examples of existing software\n",
    "\n",
    "- Commercial\n",
    "    - [GOLD](https://www.ccdc.cam.ac.uk/solutions/csd-discovery/components/gold/)\n",
    "    - [Schrödinger](https://www.schrodinger.com/glide)\n",
    "    - [FlexX](https://www.biosolveit.de/FlexX/)\n",
    "- Free (or free for academics)\n",
    "    - [AutoDock](http://autodock.scripps.edu/)\n",
    "    - [AutoDock Vina](http://vina.scripps.edu/)\n",
    "    - [DOCK](http://dock.compbio.ucsf.edu/)\n",
    "    - [OpenEye](https://www.eyesopen.com/)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparation of structures\n",
    "\n",
    "We will use the AutoDock Vina installation present in the OPAL webservices. However, this requires preparing the structures beforehand. The recommended approach is using the preparation scripts present in [AutoDockTools](http://autodock.scripps.edu/resources/adt). Although these tools are distributed on their own, and are only compatible with Python 2, we have prepared a [Python 3-ready fork](https://github.com/jaimergp/autodocktools-prepare-py3k) containing the subset needed for protein and ligand preparations. This should be enough for our needs.\n",
    "\n",
    "\n",
    "### Binding pocket prediction\n",
    "\n",
    "The docking calculation will work best if we perform it in a reasonably small search space, normally covering a single binding \n",
    "pocket. To guesstimate the best one, we can use DoGSiteScorer, available for free and online at Proteins.plus.\n",
    "\n",
    "#### Proteins.plus DoGSiteScorer\n",
    "\n",
    "* Role: Interactive web interface for several CADD tools\n",
    "* Website: http://proteins.plus\n",
    "* API: Yes, REST-based. Simple enough to apply bare `requests`\n",
    "* Documentation: https://proteins.plus/help/index\n",
    "* Literature:\n",
    "   * [_J. Chem. Inf. Model._ (2010), __50__, 11, 2041-2052](https://doi.org/10.1021/ci100241y)\n",
    "   * [_Bioinformatics_ (2012), __28__, 15, 2074–2075](https://doi.org/10.1093/bioinformatics/bts310)\n",
    "\n",
    "> Automated prediction of protein active sites is essential for large-scale protein function prediction, classification, and druggability estimates. In this work, we present DoGSite, a new structure-based method to predict active sites in proteins based on a Difference of Gaussian (DoG) approach which originates from image processing. In contrast to existing methods, DoGSite splits predicted pockets into subpockets, revealing a refined description of the topology of active sites. DoGSite correctly predicts binding pockets for over 92% of the PDBBind and the scPDB data set, being in line with the best-performing methods available. In 63% of the PDBBind data set the detected pockets can be subdivided into smaller subpockets. The cocrystallized ligand is contained in exactly one subpocket in 87% of the predictions. Furthermore, we introduce a more precise prediction performance measure by taking the pairwise ligand and pocket coverage into account. In 90% of the cases DoGSite predicts a pocket that contains at least half of the ligand. In 70% of the cases additionally more than a quarter of the respective pocket itself is covered by the cocrystallized ligand. Consideration of subpockets produces an increase in coverage yielding a success rate of 83% for the latter measure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Docking\n",
    "\n",
    "There are a couple of webservices available online for free use: SwissDock and OPAL webservices (which includes AutoDock Vina).\n",
    "\n",
    "#### SwissDock\n",
    "\n",
    "* Role: Perform docking calculations\n",
    "* Website: http://www.swissdock.ch\n",
    "* API: Yes, SOAP-based. No official client, use `suds-community`.\n",
    "* Documentation: http://www.swissdock.ch/pages/soap_access\n",
    "* Literature:\n",
    "   * [_Nucleic Acids Res._ (2011), __39__, W270-7.](https://academic.oup.com/nar/article/39/suppl_2/W270/2506492)\n",
    "   * [_J Comput Chem._ (2011), __32__, 10, 2149-59](https://onlinelibrary.wiley.com/doi/abs/10.1002/jcc.21797)\n",
    "\n",
    "> SwissDock, a web service to predict the molecular interactions that may occur between a target protein and a small molecule.\n",
    "> SwissDock is based on the docking software EADock DSS, whose algorithm consists of the following steps:\n",
    "> 1. Many binding modes are generated either in a box (local docking) or in the vicinity of all target cavities (blind docking).\n",
    "> 2. Simultaneously, their CHARMM energies are estimated on a grid.\n",
    "> 3. The binding modes with the most favorable energies are evaluated with FACTS, and clustered.\n",
    "> 4. The most favorable clusters can be visualized online and downloaded on your computer.\n",
    "\n",
    "\n",
    "#### OPAL webservices\n",
    "\n",
    "* Role: CADD as a service\n",
    "* Website: http://nbcr-222.ucsd.edu/opal2/dashboard\n",
    "* API: Yes, SOAP-based. No official client, use `suds-community`.\n",
    "* Documentation: http://nbcr-222.ucsd.edu/opal2/dashboard?command=docs (currently offline)\n",
    "* Literature:\n",
    "   * Manuscript: [_Nucleic Acids Res._ (2010), __38__, W724-31](https://academic.oup.com/nar/article/38/suppl_2/W724/1122840)\n",
    "   * Documentation: [Opal: Simple Web Services Wrappers for Scientific Applications](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.533.7960&rep=rep1&type=pdf)\n",
    "    \n",
    "> Biomedical applications have become increasingly complex, and they often require large-scale high-performance computing resources with a large number of processors and memory. The complexity of application deployment and the advances in cluster, grid and cloud computing require new modes of support for biomedical research. Scientific Software as a Service (sSaaS) enables scalable and transparent access to biomedical applications through simple standards-based Web interfaces. Towards this end, we built a production web server (http://ws.nbcr.net) in August 2007 to support the bioinformatics application called MEME. The server has grown since to include docking analysis with AutoDock and AutoDock Vina, electrostatic calculations using PDB2PQR and APBS, and off-target analysis using SMAP. All the applications on the servers are powered by Opal, a toolkit that allows users to wrap scientific applications easily as web services without any modification to the scientific codes, by writing simple XML configuration files. Opal allows both web forms-based access and programmatic access of all our applications. The Opal toolkit currently supports SOAP-based Web service access to a number of popular applications from the National Biomedical Computation Resource (NBCR) and affiliated collaborative and service projects. In addition, Opal's programmatic access capability allows our applications to be accessed through many workflow tools, including Vision, Kepler, Nimrod/K and VisTrails. From mid-August 2007 to the end of 2009, we have successfully executed 239,814 jobs. The number of successfully executed jobs more than doubled from 205 to 411 per day between 2008 and 2009. The Opal-enabled service model is useful for a wide range of applications. It provides for interoperation with other applications with Web Service interfaces, and allows application developers to focus on the scientific tool and workflow development. Web server availability: http://ws.nbcr.net."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Practice\n",
    "\n",
    "\n",
    "### Get files from part A\n",
    "\n",
    "First, we will define some variables pointing to the files we obtained in part A:\n",
    "\n",
    "- `PROTEIN`: path to the `.mol2` file containing the structure of the target, without any ligands or ions\n",
    "- `COMPLEX`: path to the `.pdb` file containing the structure of the target AND the native ligand in its binding site\n",
    "- `SMILES_FILE`: path to the `.txt` file containing the SMILES representations for all the similar compounds found in PubChem\n",
    "- `smiles`: list of the SMILES string contained in `SMILES_FILE`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROTEIN = \"data/protein.mol2\"\n",
    "COMPLEX = \"data/complex.pdb\"\n",
    "SMILES_FILE = \"data/similar_smiles.txt\"\n",
    "with open(SMILES_FILE) as f:\n",
    "    smiles = [line.strip() for line in f]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use SwissDock\n",
    "\n",
    "SwissDock uses a SOAP interface, so we will need to install `suds` for that.\n",
    "\n",
    "> Notice: SwissDock servers are not working lately. Go to the OPAL alternative below!"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "!pip install suds-community"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from suds.client import Client\n",
    "import zlib\n",
    "import string\n",
    "import requests\n",
    "\n",
    "def swissdock_client():\n",
    "    # Server seems to be down at the moment...\n",
    "    # http://swissdock.vital-it.ch/soap/ replies with 503 Unavailable\n",
    "    # because it points to wrong domain... patch it?\n",
    "    SWISSDOCK_WSDL_URL = \"http://www.swissdock.ch/soap/wsdl\"\n",
    "    r = requests.get(\"http://www.swissdock.ch/soap/wsdl\")\n",
    "    r.raise_for_status()\n",
    "    WSDL = r.text.replace(\"http://swissdock.vital-it.ch/soap/\", \"http://www.swissdock.ch/soap/\")\n",
    "    with open(\"data/swissdock.wsdl\", \"w\") as f:\n",
    "        f.write(WSDL)\n",
    "    HERE = _dh[0]\n",
    "    return Client(f\"file://{HERE}/data/swissdock.wsdl\")\n",
    "\n",
    "\n",
    "def prepare_protein(client, protein):\n",
    "    \"\"\"\n",
    "    Given a PDB file (string contents), returns PSF and CRD\n",
    "    \"\"\"\n",
    "    encoded_protein = zlib.compress(protein.encode('utf-8'))\n",
    "    job_id = client.service.prepareTarget(target=encoded_protein)\n",
    "    while True:\n",
    "        result = client.service.isTargetPrepared(jobID=job_id)\n",
    "        if result is None:\n",
    "            raise ValueError(\"No such a job present\")\n",
    "        if result in (False, \"false\", 0):\n",
    "            time.sleep(5)\n",
    "        else:  # ready!\n",
    "            break\n",
    "    protein_files = client.service.getPreparedTarget(job_id)\n",
    "    if protein_files is None or len(protein_files) != 2:\n",
    "        raise ValueError(\"Could not prepare protein!\")\n",
    "    return protein_files\n",
    "            \n",
    "\n",
    "def prepare_ligand(client, ligand):\n",
    "    \"\"\"\n",
    "    Given a MOL2 file (string contents), returns PDB, RTF, PAR.\n",
    "    \n",
    "    Ligand must be protonated beforehand!\n",
    "    \"\"\"\n",
    "    encoded_ligand = zlib.compress(ligand.encode('utf-8'))\n",
    "    job_id = client.service.prepareLigand(ligand=encoded_ligand)\n",
    "    while True:\n",
    "        result = client.service.isLigandPrepared(jobID=job_id)\n",
    "        if result is None:\n",
    "            raise ValueError(\"No such a job present\")\n",
    "        if result in (False, \"false\", 0):\n",
    "            time.sleep(5)\n",
    "        else:  # ready!\n",
    "            break\n",
    "    ligand_files = client.service.getPreparedLigand(job_id)\n",
    "    if ligand_files is None or len(ligand_files) != 3:\n",
    "        raise ValueError(\"Could not prepare ligand!\")\n",
    "    return ligand_files\n",
    "\n",
    "def dock(client, protein, ligand, name=None):\n",
    "    protein_psf, protein_crd = prepare_protein(client, protein)\n",
    "    ligand_pdb, ligand_rtf, ligand_par = prepare_ligand(client, ligand)\n",
    "    \n",
    "    if name is None:\n",
    "        name = \"teachopencadd\" + ''.join([random.choice(string.ascii_letters) for _ in range(5)])\n",
    "    job_id = client.service.startDocking(\n",
    "        protein_psf, protein_crd,\n",
    "        ligand_pdb,\n",
    "        [ligand_rtf],\n",
    "        [ligand_par],\n",
    "        name)\n",
    "    if job_id in (None, \"None\"):\n",
    "        raise ValueError(\"Docking job could not be submitted\")\n",
    "    while not client.service.isDockingTerminated(job_id):\n",
    "        time.sleep(5)\n",
    "    all_files = client.service.getPredictedDockingAllFiles(job_id)\n",
    "    with open('docking_results.zip', 'w') as f:\n",
    "        f.write(all_files)\n",
    "    target, docked = client.service.getPredictedDocking(job_id)\n",
    "    client.service.forget(job_id)\n",
    "    return target, docked\n",
    "\n",
    "def smiles_to_pdb(s, out='output.pdb'):\n",
    "    m = Chem.AddHs(Chem.MolFromSmiles(s))\n",
    "    AllChem.EmbedMolecule(m)\n",
    "    if out is None:\n",
    "        return Chem.MolToPDBBlock(m)\n",
    "    Chem.MolToPDBFile(m, out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import Mol2Writer\n",
    "except ImportError:\n",
    "    # Ugly hack to get Mol2 writer/readers in RDKit\n",
    "    import os\n",
    "    working_dir = os.getcwd()\n",
    "    os.chdir(_dh[0])\n",
    "    !wget https://raw.githubusercontent.com/rdkit/rdkit/60081d31f45fa8d5e8cef527589264c57dce7c65/rdkit/Chem/Mol2Writer.py > /dev/null\n",
    "    os.chdir(working_dir)\n",
    "    import Mol2Writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def step_03_swissdock(protein_pdb, ligand_smiles):\n",
    "    ligand = Chem.AddHs(Chem.MolFromSmiles(ligand_smiles))\n",
    "    AllChem.EmbedMolecule(ligand)\n",
    "    ligand_mol2 = Mol2Writer.MolToCommonMol2Block(ligand)\n",
    "    client = swissdock_client()\n",
    "    return dock(client, protein_pdb, ligand_mol2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> _The cell below has been disabled because SwissDock servers are currently unavailable. If you insist on running it, you will need to convert the protein mol2 to PDB first. You can use OpenBabel for that (installed in part C). Feel free to define a convenience_ `mol2_to_pdb` _function._"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# TODO: Convert from mol2 to pdb!\n",
    "# protein_pdbblock = mol2_to_pdb(PROTEIN, out=None)\n",
    "step_03_swissdock(protein_pdbblock, smiles[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform docking with OPAL webservices\n",
    "SwissDock is not working recently, so we can resort to yet another webservice. The interface is a bit more rudimentary, but it should work. However, protein and ligand must be prepared locally with `AutoDockTools`. We have prepared a Python 3-ready fork, but it's not well tested. It seems to work well enough for our purposes here, though.\n",
    "\n",
    "You can install it with:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "!pip install https://github.com/jaimergp/autodocktools-prepare-py3k/archive/master.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The protocol for the docking calculation is the following:\n",
    "\n",
    "1. Prepare the protein and the ligands with AutoDockTools (locally)\n",
    "2. Find the best possible binding pocket with [Proteins.plus' DoGSiteScorer](https://proteins.plus/2ozr#dogsite). We will use this information to configure the Vina calculation (geometric center and size of the search space)\n",
    "3. Run the Vina calculation on OPAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import os\n",
    "from io import StringIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare the structures\n",
    "\n",
    "Preparing the structure simply involves running the appropriate parts of the `AutoDockTools` library:\n",
    "\n",
    "- Reading the structure file with `MolKit`\n",
    "- Applying the correct preparer: `AD4ReceptorPreparation` for the protein, `AD4LigandPreparation` for the ligand.\n",
    "\n",
    "The preparation itself will take care of things such as:\n",
    "\n",
    "- Adding hydrogens to the protein and ligand\n",
    "- Removing strange residues that do not belong to the protein\n",
    "- Assigning atom types and partial charges\n",
    "- Recognizing torsionable branches in the ligand\n",
    "\n",
    "The results are written to disk as `PDBQT` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "#\n",
    "# Structure preparation\n",
    "#\n",
    "######################\n",
    "\n",
    "import MolKit\n",
    "from AutoDockTools.MoleculePreparation import AD4ReceptorPreparation, AD4LigandPreparation\n",
    "\n",
    "def opal_prepare_protein(protein):\n",
    "    \"\"\"\n",
    "    AutoDock expects PDBQT files\n",
    "    \"\"\"\n",
    "    mol = MolKit.Read(protein)[0]\n",
    "    mol.buildBondsByDistance()\n",
    "    RPO = AD4ReceptorPreparation(mol, outputfilename=protein+'.pdbqt')\n",
    "    return protein + '.pdbqt'\n",
    "\n",
    "def opal_prepare_ligand(ligand):\n",
    "    \"\"\"\n",
    "    AutoDock expects PDBQT files\n",
    "    \"\"\"\n",
    "    mol = MolKit.Read(ligand)[0]\n",
    "    mol.buildBondsByDistance()\n",
    "    RPO = AD4LigandPreparation(mol, outputfilename=ligand+'.pdbqt')\n",
    "    return ligand + '.pdbqt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Guess the binding sites\n",
    "\n",
    "The DoGSiteScorer installation present in Proteins.plus provides a REST API that is very easy to use, if you only need to process a protein present in the PDB database (see `dogsite_scorer_submit_with_pdbid`). However, since we are using structures provided by KLIFS, we cannot guarantee that the position and orientation of the structure present in the official PDB registries match the one present in KLIFS. Although we could superpose both and then apply the resulting transformation matrix to the obtained pockets, it would be easier to simply upload our PDB file to Proteins.plus, as provided in the standard web interface.\n",
    "\n",
    "However, the REST API does not provide such an option, so we have to reverse engineer the process. To locate the appropriate HTTP requests, you would need to open the Network tab in the Chrome Developer Tools and start writing down which requests were performed as you use the website normally. Authenticity tokens and HTTP headers are key to obtain a valid request here!\n",
    "\n",
    "The results of this research are consolidated in the `dogsite_scorer_submit_with_custom_pdb` function. If you are interested in the technical details, read through the comments in the function. \n",
    "\n",
    "> For this approach, you will also need `BeautifulSoup` to parse some HTML code during the binding site guessing:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "!pip install BeautifulSoup4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this method we can now obtain the geometric center of the most probable binding pocket, as well as its size. Both values are needed to configure the Vina calculation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "#\n",
    "# Guess binding pocket\n",
    "#\n",
    "######################\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "\n",
    "def dogsite_scorer_submit_with_pdbid(pdbid, chain):\n",
    "    \"\"\"\n",
    "    This is the official API, but they only allow PDB codes\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    pdbid : str\n",
    "        4-letter PDB identifier\n",
    "    chain : str\n",
    "        Structure chain to be analyzed\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        URL that can be queried to get updates on whether the job is running or finished\n",
    "    \"\"\"\n",
    "    # Submit job to Proteins.plus\n",
    "    r = requests.post(\"https://proteins.plus/api/dogsite_rest\",\n",
    "        json={\n",
    "            \"dogsite\": {\n",
    "                \"pdbCode\": pdbid,\n",
    "                \"analysisDetail\": \"1\",\n",
    "                \"bindingSitePredictionGranularity\": \"1\",\n",
    "                \"ligand\": \"\",\n",
    "                \"chain\": chain\n",
    "            }\n",
    "        },\n",
    "        headers= {'Content-type': 'application/json', 'Accept': 'application/json'}\n",
    "    )\n",
    "\n",
    "    r.raise_for_status()\n",
    "    # We have to query location for updates on the calculation\n",
    "    return r.json()['location']\n",
    "\n",
    "def dogsite_scorer_submit_with_custom_pdb(pdbfile):\n",
    "    \"\"\"\n",
    "    In order to upload a custom PDB, we have to mimic the actual HTML frontend:\n",
    "    \n",
    "    1. Obtain the CSRF token out the HTML meta headers\n",
    "    2. Post the file to upload it\n",
    "    3. The returned HTML page will contain the URL ID, which in turn will allow\n",
    "       us to obtain the internal shorthand job ID. We can use that one to\n",
    "       retrieve the public job API ID by mimicking the async calls that the\n",
    "       webserver performs in the frontend (as if we were using the web interface).\n",
    "    4. Once we obtain the public job ID we can switch to using the REST API.\n",
    "    \"\"\"\n",
    "    # We need to use a `session` to store intermediate cookies during the process\n",
    "    session = requests.Session()\n",
    "    r = session.get(\"https://proteins.plus/\")\n",
    "    r.raise_for_status()\n",
    "    # The homepage contains the CSRF token needed to validate our request\n",
    "    # Otherwise it wouldn't be safe! We have to use that throughout our requests\n",
    "    # so the best way is to set it as part of the session HTTP headers\n",
    "    html = BeautifulSoup(r.text)\n",
    "    token = html.find('input', {'name': 'authenticity_token'}).attrs['value']\n",
    "    session.headers['X-CSRF-Token'] = token\n",
    "\n",
    "    # 1. Upload file\n",
    "    with open(pdbfile, 'rb') as f:\n",
    "        r = session.post(\"https://proteins.plus\", files={'pdb_file[pathvar]': f})\n",
    "    r.raise_for_status()\n",
    "\n",
    "    # If the REST API supported file uploads, we would have the public ID already\n",
    "    # but in the meantime you will have to work around it this way\n",
    "        \n",
    "    # 2. Get internal location id\n",
    "    html = BeautifulSoup(r.text)\n",
    "    pdb_id = html.find('input', {'name': \"dogsite[pdbCode]\"}).attrs['value']\n",
    "\n",
    "    # 3. Get the internal job ID\n",
    "    session.headers['Referer'] = \"https://proteins.plus\" + pdb_id\n",
    "    r = session.post(f\"https://proteins.plus/{pdb_id}/dogsites\",\n",
    "            json={\"dogsite\": {\"pdbCode\": pdb_id}},\n",
    "            headers= {'Content-type': 'application/json', \n",
    "                      'Accept': 'application/json'}\n",
    "    )\n",
    "    r.raise_for_status()\n",
    "    job_id = r.json()['job_id']\n",
    "    time.sleep(3)  # wait a bit before continuing so the server can process the request\n",
    "    \n",
    "    # 4. Get the public job ID\n",
    "    while True:\n",
    "        r = session.get(f\"https://proteins.plus/{pdb_id}/dogsites/{job_id}?_={round(time.time())}\",\n",
    "                        headers= {\n",
    "                            'Accept': 'application/json, text/javascript, */*',\n",
    "                            'Sec-Fetch-Mode': 'cors',\n",
    "                            'Sec-Fetch-Site': 'same-origin',\n",
    "                            # this line below makes all the difference, apparently\n",
    "                            # otherwise, error 406 is thrown\n",
    "                            'X-Requested-With': 'XMLHttpRequest'}\n",
    "                       )\n",
    "        r.raise_for_status()\n",
    "        if 'Calculation in progress...' in r.text:  # not finished yet\n",
    "            time.sleep(5)\n",
    "            continue\n",
    "        if 'Error during DogSiteScorer calculation' in r.text:  # malformed file?\n",
    "            raise ValueError('Could not run DoGSiteScorer!')\n",
    "        break\n",
    "    \n",
    "    results_id = None\n",
    "    for lines in r.text.splitlines():\n",
    "        for line in lines.split('\\\\n'):\n",
    "            if 'results/dogsite' in line:\n",
    "                results_id = line.split('/')[3]\n",
    "                break\n",
    "    if results_id is None:\n",
    "        raise ValueError(r.text)\n",
    "        \n",
    "    return f\"https://proteins.plus/api/dogsite_rest/{results_id}\"\n",
    "    \n",
    "\n",
    "def dogsite_scorer_guess_binding_site(protein):\n",
    "    \"\"\"\n",
    "    Use Proteins.plus' DoGSiteScorer to retrieve most probable binding site in protein.\n",
    "    \"\"\"\n",
    "    if len(protein) == 4:  # pdb code\n",
    "        job_location = dogsite_scorer_submit_with_pdbid(protein)\n",
    "    elif protein.endswith('.pdb'):\n",
    "        job_location = dogsite_scorer_submit_with_custom_pdb(protein)\n",
    "    else:\n",
    "        raise ValueError(\"`protein` must be a PDB ID or a path to a .pdb file!\")\n",
    "    \n",
    "    # Check when the calculation has finished\n",
    "    while True:\n",
    "        result = requests.get(job_location)\n",
    "        result.raise_for_status()  # if it fails, it will stop here\n",
    "        if result.status_code == 202:  # still running\n",
    "            time.sleep(5)\n",
    "            continue\n",
    "        break\n",
    "    \n",
    "    # The residues files contain the geometric center and radius as a comment in the PDB file\n",
    "    # first file (residues[0]) is the best scored pocket\n",
    "    pdb_residues = requests.get(result.json()['residues'][0]).text\n",
    "    for line in pdb_residues.splitlines():\n",
    "        line = line.strip()\n",
    "        if line.startswith('HEADER') and 'Geometric pocket center at' in line:\n",
    "            fields = line.split()\n",
    "            center = [float(x) for x in fields[5:8]]\n",
    "            radius = float(fields[-1])\n",
    "            break\n",
    "    return center, radius  # this is what we need for our Vina calculation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run Vina on OPAL\n",
    "\n",
    "\n",
    "Once we have (1) prepared the protein and ligand, and (2) guessed the search area, we can submit the actual calculation to the OPAL web servers. This involves:\n",
    "\n",
    "1. Initializing the SOAP client with `suds`\n",
    "2. Encoding the files as [`base64` strings](https://en.wikipedia.org/wiki/Base64#Examples) so they can be attached to the SOAP XML request\n",
    "3. Submitting the job request and getting the job ID back\n",
    "4. Querying the server with the job ID to check whether the calculation is finished or not\n",
    "5. Downloading the relevant output files with `requests`\n",
    "\n",
    "> These steps are not very well documented in the OPAL website (actually, the documentation is [unavailable](http://rocce-vm0.ucsd.edu/data/docs/opal/documentation.html)), so they were figured out by exploring the source code present in some of the [UCSF Chimera modules](http://plato.cgl.ucsf.edu/trac/chimera/browser/trunk/libs/WebServices/opal_client.py) that rely on these servers.\n",
    "\n",
    "Since the calculation usually takes 5-15 minutes, for the 4th step we will use some of the Jupyter goodies and update the contents of the output file in realtime (see function `iprint()`). That way, we can check the progress instead of blindly trusting that the calculation is running ok!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "#\n",
    "# Run calculation\n",
    "#\n",
    "######################\n",
    "\n",
    "from suds.client import Client\n",
    "from IPython.display import display, clear_output, HTML\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import AllChem\n",
    "\n",
    "VINA_CONFIG = \"\"\"\n",
    "center_x = {center[0]}\n",
    "center_y = {center[1]}\n",
    "center_z = {center[2]}\n",
    "size_x = {size[0]}\n",
    "size_y = {size[1]}\n",
    "size_z = {size[2]}\n",
    "\"\"\"\n",
    "\n",
    "def opal_run_docking(protein, ligand, center, size, stream_output=True):\n",
    "    \"\"\"\n",
    "    Connect to OPAL webservices and submit job\n",
    "    \"\"\"\n",
    "    client = Client(\"http://nbcr-222.ucsd.edu/opal2/services/vina_1.1.2?wsdl\")\n",
    "    files = 'receptor.pdbqt', 'ligand.pdbqt', 'vina.conf'\n",
    "    with open(protein) as f:\n",
    "        protein_contents = f.read()\n",
    "    with open(ligand) as f:\n",
    "        ligand_contents = f.read()\n",
    "    file_map = [\n",
    "        {'name': 'receptor.pdbqt',\n",
    "         'contents': base64ify(protein_contents)},\n",
    "        {'name': 'ligand.pdbqt',\n",
    "         'contents': base64ify(ligand_contents)},\n",
    "        {'name': 'vina.conf',\n",
    "         'contents': base64ify(VINA_CONFIG.format(center=center, size=size))},\n",
    "        {'name': 'results.pdbqt',\n",
    "         'contents': ''},\n",
    "    ]\n",
    "    cli_args = \"--receptor receptor.pdbqt --ligand ligand.pdbqt --config vina.conf --out results.pdbqt\"\n",
    "    \n",
    "    response = client.service.launchJob(cli_args, inputFile=file_map)\n",
    "    job_id = response.jobID\n",
    "    url = f\"http://nbcr-222.ucsd.edu/opal-jobs/{job_id}\"\n",
    "    message = \"Waiting for job \" + url\n",
    "    while True:\n",
    "        r = requests.get(url + \"/vina.out\")\n",
    "        try:\n",
    "            r.raise_for_status()\n",
    "        except:  # output file might not exist yet during the first checks\n",
    "            iprint(message)\n",
    "        else:\n",
    "            iprint(f\"{message}\\n{r.text}\")\n",
    "        if client.service.queryStatus(job_id).code == 2:\n",
    "            time.sleep(10)\n",
    "            continue\n",
    "        print('\\nFinished!')\n",
    "        break\n",
    "        \n",
    "    output_response = client.service.getOutputs(job_id)\n",
    "    output_files = {\n",
    "        'stdout.txt': requests.get(output_response.stdOut).text,\n",
    "        'stderr.txt': requests.get(output_response.stdErr).text,\n",
    "    }\n",
    "    for f in output_response.outputFile:\n",
    "        if f.name in files:\n",
    "            continue\n",
    "        r = requests.get(f.url)\n",
    "        r.encoding = 'utf-8'\n",
    "        r.raise_for_status()\n",
    "        contents = r.text\n",
    "        output_files[f.name] = contents \n",
    "        time.sleep(0.1)\n",
    "    \n",
    "    return output_files\n",
    "\n",
    "######################\n",
    "#\n",
    "# Utilities\n",
    "#\n",
    "######################\n",
    "\n",
    "import base64\n",
    "\n",
    "def base64ify(bytes_or_str):\n",
    "    \"\"\"\n",
    "    Mimic Py2k base64encode behavior\n",
    "    \"\"\"\n",
    "    if isinstance(bytes_or_str, str):\n",
    "        input_bytes = bytes_or_str.encode('utf8')\n",
    "    else:\n",
    "        input_bytes = bytes_or_str\n",
    "\n",
    "    output_bytes = base64.urlsafe_b64encode(input_bytes)\n",
    "    return output_bytes.decode('ascii')\n",
    "\n",
    "def iprint(s):\n",
    "    \"\"\"\n",
    "    We can use this function to print outputs, overwriting previous ones, so it\n",
    "    looks like it's constantly updating :)\n",
    "    \"\"\"\n",
    "    clear_output(wait=True)\n",
    "    s = s.replace(\"\\n\", \"<br />\")\n",
    "    display(HTML(f'<pre>{s}</pre>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Put it all together\n",
    "\n",
    "Now that all the needed functions are defined, we can create the pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def step_03_opal(protein, smiles, pdbcomplex):\n",
    "    \"\"\"\n",
    "    Given a protein structure and a list of smiles strings:\n",
    "    Steps:\n",
    "        1. Prepare the protein for AutoDock Vina (locally)\n",
    "        2. Use DoGSiteScorer to find the most probable binding site\n",
    "        3. For each ligand, use RDKit to write a 3D PDB file and\n",
    "           run AutoDockVina on OPAL\n",
    "    \n",
    "    The whole thing should take around 5-15 mins\n",
    "    \n",
    "    The result is a dictionary with the output file contents. We \n",
    "    are mainly interested in result['results.pdbqt']\n",
    "    \"\"\"\n",
    "    prepared_protein = opal_prepare_protein(protein)\n",
    "    center, radius = dogsite_scorer_guess_binding_site(pdbcomplex)\n",
    "    size = [radius] * 3  # Vina supports non-cubic boxes, but we will use a cube for simplicity\n",
    "    for i, smile in enumerate(smiles):\n",
    "        smiles_to_pdb(smile, f'data/ligand{i}.pdb')\n",
    "        prepared_ligand = opal_prepare_ligand(f'data/ligand{i}.pdb')\n",
    "        result = opal_run_docking(prepared_protein, prepared_ligand, center, size)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run it! `%time` magic will measure how long it takes for us."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>Waiting for job http://nbcr-222.ucsd.edu/opal-jobs/appvina_1.1.21573834866204641963381<br />#################################################################<br /># If you used AutoDock Vina in your work, please cite:          #<br />#                                                               #<br /># O. Trott, A. J. Olson,                                        #<br /># AutoDock Vina: improving the speed and accuracy of docking    #<br /># with a new scoring function, efficient optimization and       #<br /># multithreading, Journal of Computational Chemistry 31 (2010)  #<br /># 455-461                                                       #<br />#                                                               #<br /># DOI 10.1002/jcc.21334                                         #<br />#                                                               #<br /># Please see http://vina.scripps.edu for more information.      #<br />#################################################################<br /><br />Reading input ... done.<br />Setting up the scoring function ... done.<br />Analyzing the binding site ... done.<br />Using random seed: 1064637608<br />Performing search ... <br />0%   10   20   30   40   50   60   70   80   90   100%<br />|----|----|----|----|----|----|----|----|----|----|<br />***************************************************<br />done.<br />Refining results ... done.<br /><br />mode |   affinity | dist from best mode<br />     | (kcal/mol) | rmsd l.b.| rmsd u.b.<br />-----+------------+----------+----------<br />   1        -10.4      0.000      0.000<br />   2        -10.2      1.798      2.851<br />   3         -9.9      3.942      9.366<br />   4         -9.2      1.689      2.122<br />   5         -9.2      4.645     10.800<br />   6         -8.7      2.749      3.527<br />   7         -8.7      4.620     10.696<br />   8         -8.5      2.792      5.325<br />   9         -8.0      3.208      6.296<br />Writing output ... done.<br /></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Finished!\n",
      "CPU times: user 2.99 s, sys: 165 ms, total: 3.15 s\n",
      "Wall time: 9min 40s\n"
     ]
    }
   ],
   "source": [
    "# We will only process the first ligand in `smiles`\n",
    "%time result = step_03_opal(PROTEIN, smiles[:1], COMPLEX)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Understanding the output\n",
    "\n",
    "`result` is a dictionary with several keys, corresponding to the output files and their text contents. We are mainly interested in:\n",
    "\n",
    "- `results.pdbqt` contains the docked ligands. It's a modified multi-model PDB file. Since we kept the protein rigid, we just need to open each ligand model together with the original protein structure.\n",
    "- `vina.out` is the text output you see above. It will provide the table-like information.\n",
    "\n",
    "Save them to disk so we can retrieve them later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/results.pdbqt', 'w') as f:\n",
    "    f.write(result['results.pdbqt'])\n",
    "with open('data/vina.out', 'w') as f:\n",
    "    f.write(result['vina.out'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualize docking results\n",
    "\n",
    "Once the calculation has run and the files have been downloaded, it's time to visualize them! You will see how to do that in part C.\n",
    "\n",
    "## Discussion\n",
    "\n",
    "OPAL offers a Vina installation to perform docking calculations for free, but we have to prepare the input files locally before the submission. Part of that preparation involves defining the search area, normally around a known binding site. Instead of guessing visually, we have used the DoGSiteScorer server present in Proteins.plus to calculate the origin and radius of the most probable binding site. These two services used different communication interfaces.\n",
    "\n",
    "Proteins.plus uses REST, but does not provide a `swagger.json` definition as KLIFS did, so we had to build our own requests manually. Fortunately, the service is simple enough to *only* need a couple of requests... or so we thought! The current API does not allow for custom PDB uploads, so we had to use a handful of GET and POST requests to pretend we were using a normal browser. To guess the correct requests, the Network tab in Chrome Developer Tools was really handy. If you ever need to reverse engineer a webservice, this is one of the tools you can use!\n",
    "\n",
    "OPAL is built with the XML-based SOAP standard, a bit more cumbersome than JSON-based REST, but `suds` makes it really easy. Although `suds` will report the available methods, these are not very well documented. Fortunately, there are some code examples scattered in some modules of [UCSF Chimera](http://plato.cgl.ucsf.edu/trac/chimera/browser/trunk/libs/WebServices/opal_client.py). Jobs submitted to OPAL are public once you have the job ID and the files are updated in realtime, so we could build a live preview of the calculation output by querying the server every N seconds and updating the `display()`ed HTML dynamically. This trick can also be used to query a calculation happening in your machine: all you need to do is read the file contents, clear the current output and display the new one as an `HTML()` object (see function `iprint()`).\n",
    "\n",
    "\n",
    "## Quiz\n",
    "\n",
    "- How can you tell if the docking has run successfully in the remote server?\n",
    "- Why do we need to prepare the AutoDock Vina input files locally?\n",
    "- Advanced: Can you find the correct HTTP requests to use [MCule's docking server](https://mcule.com/apps/1-click-docking/) from the Notebook?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
